{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import gzip\n",
    "import io\n",
    "from scripts.check_overlaps import check_overlaps, check_overlaps_and_annotate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rheinbay 2020 Supplementary Table 5<br>\n",
    "We filter on the basis of 'Final filter judgement' = TRUE<br>\n",
    "Then we take the unique element IDs<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel('data/Rheinbay 2020/41586_2020_1965_MOESM4_ESM.xlsx', engine='openpyxl', sheet_name='Table 5 Non-coding point mut dr')\n",
    "# df1 = pd.read_excel('data/Rheinbay 2020/41586_2020_1965_MOESM4_ESM.xlsx', engine='openpyxl', sheet_name='Table 4 Protein-coding point mu')\n",
    "df['file_name'] = df['ID'].apply(lambda x: x.split('::')[0]+'.bed')\n",
    "df['file_name'] = df['file_name'].apply(lambda x: x.replace('mirna_', 'mirna.') if 'mirna_pre' in x or 'mirna_mat' in x else x)\n",
    "noncoding_elements = list(df[df['Final filter judgement'] == True]['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_df = pd.read_csv('data/PCAWG_mutations_to_elements.icgc.public.txt', sep='\\t')\n",
    "file_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gc19_pc.cds::gencode::TP53::ENSG00000141510.11           516\n",
       "gc19_pc.cds::gencode::TTN::ENSG00000155657.19            476\n",
       "lncrna.ncrna::gencode::SNHG14::ENSG00000224078.8         233\n",
       "gc19_pc.cds::gencode::KRAS::ENSG00000133703.7            223\n",
       "lncrna.ncrna::gencode::KCNQ1OT1::ENSG00000269821.1       221\n",
       "                                                        ... \n",
       "gc19_pc.5utr::gencode::GRIK1-AS2::ENSG00000183653.4        1\n",
       "gc19_pc.5utr::gencode::C5orf46::ENSG00000178776.4          1\n",
       "gc19_pc.5utr::gencode::ZNF184::ENSG00000096654.11          1\n",
       "gc19_pc.5utr::gencode::ZNF853::ENSG00000236609.3           1\n",
       "lncrna.ncrna::gencode::RP11-7O11.3::ENSG00000237950.1      1\n",
       "Name: reg_id, Length: 108242, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_df['reg_id'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df['Final filter judgement'] == True]['ID'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>region_type</th>\n",
       "      <th>reg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>111231125</td>\n",
       "      <td>111231125</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11</td>\n",
       "      <td>111232885</td>\n",
       "      <td>111232885</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>86866372</td>\n",
       "      <td>86866372</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr7:86865600-86866400::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2267139</td>\n",
       "      <td>2267139</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr1:2267000-2269600::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11</td>\n",
       "      <td>111226848</td>\n",
       "      <td>111226848</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901</th>\n",
       "      <td>11</td>\n",
       "      <td>65194294</td>\n",
       "      <td>65194295</td>\n",
       "      <td>-</td>\n",
       "      <td>CATAATATTTTAAAA</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "      <td>lncrna.ncrna::gencode::NEAT1::ENSG00000245532.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>902</th>\n",
       "      <td>11</td>\n",
       "      <td>65269272</td>\n",
       "      <td>65269273</td>\n",
       "      <td>AA</td>\n",
       "      <td>-</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "      <td>lncrna.ncrna::gencode::MALAT1::ENSG00000251562.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903</th>\n",
       "      <td>11</td>\n",
       "      <td>65267659</td>\n",
       "      <td>65267659</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "      <td>lncrna.ncrna::gencode::MALAT1::ENSG00000251562.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904</th>\n",
       "      <td>11</td>\n",
       "      <td>65209144</td>\n",
       "      <td>65209148</td>\n",
       "      <td>TTTAT</td>\n",
       "      <td>-</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "      <td>lncrna.ncrna::gencode::NEAT1::ENSG00000245532.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>11</td>\n",
       "      <td>65196438</td>\n",
       "      <td>65196454</td>\n",
       "      <td>ACATTTTTAAACAAAAA</td>\n",
       "      <td>-</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "      <td>lncrna.ncrna::gencode::NEAT1::ENSG00000245532.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>906 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr      start        end                ref              alt  \\\n",
       "0    11  111231125  111231125                  G                C   \n",
       "1    11  111232885  111232885                  C                T   \n",
       "2     7   86866372   86866372                  T                C   \n",
       "3     1    2267139    2267139                  G                A   \n",
       "4    11  111226848  111226848                  T                C   \n",
       "..   ..        ...        ...                ...              ...   \n",
       "901  11   65194294   65194295                  -  CATAATATTTTAAAA   \n",
       "902  11   65269272   65269273                 AA                -   \n",
       "903  11   65267659   65267659                  A                -   \n",
       "904  11   65209144   65209148              TTTAT                -   \n",
       "905  11   65196438   65196454  ACATTTTTAAACAAAAA                -   \n",
       "\n",
       "          region_type                                            reg_id  \n",
       "0       enhancers.bed      enhancers::chr11:111222000-111233000::NA::NA  \n",
       "1       enhancers.bed      enhancers::chr11:111222000-111233000::NA::NA  \n",
       "2       enhancers.bed         enhancers::chr7:86865600-86866400::NA::NA  \n",
       "3       enhancers.bed           enhancers::chr1:2267000-2269600::NA::NA  \n",
       "4       enhancers.bed      enhancers::chr11:111222000-111233000::NA::NA  \n",
       "..                ...                                               ...  \n",
       "901  lncrna.ncrna.bed   lncrna.ncrna::gencode::NEAT1::ENSG00000245532.4  \n",
       "902  lncrna.ncrna.bed  lncrna.ncrna::gencode::MALAT1::ENSG00000251562.3  \n",
       "903  lncrna.ncrna.bed  lncrna.ncrna::gencode::MALAT1::ENSG00000251562.3  \n",
       "904  lncrna.ncrna.bed   lncrna.ncrna::gencode::NEAT1::ENSG00000245532.4  \n",
       "905  lncrna.ncrna.bed   lncrna.ncrna::gencode::NEAT1::ENSG00000245532.4  \n",
       "\n",
       "[906 rows x 7 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rheinbay_noncoding_drivers = file_df[file_df['reg_id'].isin(noncoding_elements)][['mut_chr', 'mut_pos1', 'mut_pos2', 'mut_ref', 'mut_alt', 'region_type', 'reg_id']].drop_duplicates().reset_index(drop=True)\n",
    "rheinbay_noncoding_drivers.rename({'mut_chr': 'chr', 'mut_pos1': 'start', 'mut_pos2': 'end', 'mut_ref': 'ref', 'mut_alt': 'alt'}, axis=1, inplace=True)\n",
    "rheinbay_noncoding_drivers['chr'] = rheinbay_noncoding_drivers['chr'].apply(lambda x: x.replace('chr', ''))\n",
    "rheinbay_noncoding_drivers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>region_type</th>\n",
       "      <th>reg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>chr5</td>\n",
       "      <td>1295228</td>\n",
       "      <td>1295228</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>gc19_pc.promCore::gencode::TERT::ENSG000001643...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>chr5</td>\n",
       "      <td>1295161</td>\n",
       "      <td>1295161</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>gc19_pc.promCore::gencode::TERT::ENSG000001643...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>chr5</td>\n",
       "      <td>1295133</td>\n",
       "      <td>1295133</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>gc19_pc.promCore::gencode::TERT::ENSG000001643...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>chr5</td>\n",
       "      <td>1295228</td>\n",
       "      <td>1295228</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>gc19_pc.promCore::gencode::TERT::ENSG000001643...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr    start      end ref alt           region_type  \\\n",
       "261  chr5  1295228  1295228   G   A  gc19_pc.promCore.bed   \n",
       "300  chr5  1295161  1295161   T   G  gc19_pc.promCore.bed   \n",
       "389  chr5  1295133  1295133   C   T  gc19_pc.promCore.bed   \n",
       "421  chr5  1295228  1295228   G   T  gc19_pc.promCore.bed   \n",
       "\n",
       "                                                reg_id  \n",
       "261  gc19_pc.promCore::gencode::TERT::ENSG000001643...  \n",
       "300  gc19_pc.promCore::gencode::TERT::ENSG000001643...  \n",
       "389  gc19_pc.promCore::gencode::TERT::ENSG000001643...  \n",
       "421  gc19_pc.promCore::gencode::TERT::ENSG000001643...  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rheinbay_noncoding_drivers[rheinbay_noncoding_drivers['reg_id'].str.contains('TERT')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>region_type</th>\n",
       "      <th>reg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>19578253</td>\n",
       "      <td>19578254</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MRTO4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>19578254</td>\n",
       "      <td>19578255</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MRTO4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>23889291</td>\n",
       "      <td>23889292</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Ovary</td>\n",
       "      <td>ID3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>23895459</td>\n",
       "      <td>23895460</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Ovary</td>\n",
       "      <td>ID3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>23888768</td>\n",
       "      <td>23888769</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Ovary</td>\n",
       "      <td>ID3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>9</td>\n",
       "      <td>125026995</td>\n",
       "      <td>125026996</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Breast</td>\n",
       "      <td>MRRF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>9</td>\n",
       "      <td>124049461</td>\n",
       "      <td>124049462</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Breast</td>\n",
       "      <td>MRRF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>X</td>\n",
       "      <td>70338403</td>\n",
       "      <td>70338404</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>X</td>\n",
       "      <td>70401837</td>\n",
       "      <td>70401838</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>X</td>\n",
       "      <td>70365659</td>\n",
       "      <td>70365660</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>215 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr      start        end ref alt region_type reg_id\n",
       "0     1   19578253   19578254   G   A       Brain  MRTO4\n",
       "1     1   19578254   19578255   G   A       Brain  MRTO4\n",
       "2     1   23889291   23889292   G   T       Ovary    ID3\n",
       "3     1   23895459   23895460   G   T       Ovary    ID3\n",
       "4     1   23888768   23888769   T   C       Ovary    ID3\n",
       "..   ..        ...        ...  ..  ..         ...    ...\n",
       "210   9  125026995  125026996   C   G      Breast   MRRF\n",
       "211   9  124049461  124049462   T   A      Breast   MRRF\n",
       "212   X   70338403   70338404   G   A       Brain  MED12\n",
       "213   X   70401837   70401838   G   T       Brain  MED12\n",
       "214   X   70365659   70365660   G   A       Brain  MED12\n",
       "\n",
       "[215 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dr_nod = pd.read_excel('data/Dr. Nod/gkac1251_supplemental_files/SupplementaryTable4.xlsx', engine='openpyxl')\n",
    "dr_nod = dr_nod[['chr', 'pos0', 'pos1', 'ref', 'alt', 'tissue', 'geneSymbol']]\n",
    "dr_nod.rename({'pos0': 'start', 'pos1': 'end', 'tissue': 'region_type', 'geneSymbol': 'reg_id'}, inplace=True, axis=1)\n",
    "dr_nod['chr'] = dr_nod['chr'].apply(lambda x: x.replace('chr', ''))\n",
    "dr_nod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>region_type</th>\n",
       "      <th>reg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>111231125</td>\n",
       "      <td>111231125</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11</td>\n",
       "      <td>111232885</td>\n",
       "      <td>111232885</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>86866372</td>\n",
       "      <td>86866372</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr7:86865600-86866400::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2267139</td>\n",
       "      <td>2267139</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr1:2267000-2269600::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11</td>\n",
       "      <td>111226848</td>\n",
       "      <td>111226848</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>9</td>\n",
       "      <td>125026995</td>\n",
       "      <td>125026996</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Breast</td>\n",
       "      <td>MRRF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>9</td>\n",
       "      <td>124049461</td>\n",
       "      <td>124049462</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Breast</td>\n",
       "      <td>MRRF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>X</td>\n",
       "      <td>70338403</td>\n",
       "      <td>70338404</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>X</td>\n",
       "      <td>70401837</td>\n",
       "      <td>70401838</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>X</td>\n",
       "      <td>70365659</td>\n",
       "      <td>70365660</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1121 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr      start        end ref alt    region_type  \\\n",
       "0    11  111231125  111231125   G   C  enhancers.bed   \n",
       "1    11  111232885  111232885   C   T  enhancers.bed   \n",
       "2     7   86866372   86866372   T   C  enhancers.bed   \n",
       "3     1    2267139    2267139   G   A  enhancers.bed   \n",
       "4    11  111226848  111226848   T   C  enhancers.bed   \n",
       "..   ..        ...        ...  ..  ..            ...   \n",
       "210   9  125026995  125026996   C   G         Breast   \n",
       "211   9  124049461  124049462   T   A         Breast   \n",
       "212   X   70338403   70338404   G   A          Brain   \n",
       "213   X   70401837   70401838   G   T          Brain   \n",
       "214   X   70365659   70365660   G   A          Brain   \n",
       "\n",
       "                                           reg_id  \n",
       "0    enhancers::chr11:111222000-111233000::NA::NA  \n",
       "1    enhancers::chr11:111222000-111233000::NA::NA  \n",
       "2       enhancers::chr7:86865600-86866400::NA::NA  \n",
       "3         enhancers::chr1:2267000-2269600::NA::NA  \n",
       "4    enhancers::chr11:111222000-111233000::NA::NA  \n",
       "..                                            ...  \n",
       "210                                          MRRF  \n",
       "211                                          MRRF  \n",
       "212                                         MED12  \n",
       "213                                         MED12  \n",
       "214                                         MED12  \n",
       "\n",
       "[1121 rows x 7 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data = pd.concat([rheinbay_noncoding_drivers, dr_nod])\n",
    "test_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>region_type</th>\n",
       "      <th>reg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>111231125</td>\n",
       "      <td>111231125</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11</td>\n",
       "      <td>111232885</td>\n",
       "      <td>111232885</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>86866372</td>\n",
       "      <td>86866372</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr7:86865600-86866400::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2267139</td>\n",
       "      <td>2267139</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr1:2267000-2269600::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11</td>\n",
       "      <td>111226848</td>\n",
       "      <td>111226848</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057</th>\n",
       "      <td>9</td>\n",
       "      <td>125026995</td>\n",
       "      <td>125026996</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Breast</td>\n",
       "      <td>MRRF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1058</th>\n",
       "      <td>9</td>\n",
       "      <td>124049461</td>\n",
       "      <td>124049462</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Breast</td>\n",
       "      <td>MRRF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1059</th>\n",
       "      <td>X</td>\n",
       "      <td>70338403</td>\n",
       "      <td>70338404</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060</th>\n",
       "      <td>X</td>\n",
       "      <td>70401837</td>\n",
       "      <td>70401838</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1061</th>\n",
       "      <td>X</td>\n",
       "      <td>70365659</td>\n",
       "      <td>70365660</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1062 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr      start        end ref alt    region_type  \\\n",
       "0     11  111231125  111231125   G   C  enhancers.bed   \n",
       "1     11  111232885  111232885   C   T  enhancers.bed   \n",
       "2      7   86866372   86866372   T   C  enhancers.bed   \n",
       "3      1    2267139    2267139   G   A  enhancers.bed   \n",
       "4     11  111226848  111226848   T   C  enhancers.bed   \n",
       "...   ..        ...        ...  ..  ..            ...   \n",
       "1057   9  125026995  125026996   C   G         Breast   \n",
       "1058   9  124049461  124049462   T   A         Breast   \n",
       "1059   X   70338403   70338404   G   A          Brain   \n",
       "1060   X   70401837   70401838   G   T          Brain   \n",
       "1061   X   70365659   70365660   G   A          Brain   \n",
       "\n",
       "                                            reg_id  \n",
       "0     enhancers::chr11:111222000-111233000::NA::NA  \n",
       "1     enhancers::chr11:111222000-111233000::NA::NA  \n",
       "2        enhancers::chr7:86865600-86866400::NA::NA  \n",
       "3          enhancers::chr1:2267000-2269600::NA::NA  \n",
       "4     enhancers::chr11:111222000-111233000::NA::NA  \n",
       "...                                            ...  \n",
       "1057                                          MRRF  \n",
       "1058                                          MRRF  \n",
       "1059                                         MED12  \n",
       "1060                                         MED12  \n",
       "1061                                         MED12  \n",
       "\n",
       "[1062 rows x 7 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data = test_data.drop_duplicates(subset=['chr', 'start', 'end', 'ref', 'alt']).reset_index(drop=True)\n",
    "test_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_data['driver'] = 1\n",
    "test_data.to_csv('data/test_data_final.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# elements_list = os.listdir('data/gene-coding-and-non-coding-elements/')\n",
    "# elements_list.remove('allCombined.zip')\n",
    "# elements_list.remove('mapFiles')\n",
    "\n",
    "# arr = []\n",
    "# df['chr'] = ''\n",
    "# df['start'] = 0\n",
    "# df['end'] = 0\n",
    "# for element in elements_list:\n",
    "#     print(element)\n",
    "#     tempdf = df[df['file_name'] == element]\n",
    "#     file_df = pd.read_csv('data/gene-coding-and-non-coding-elements/'+element, sep='\\t', header=None)\n",
    "#     file_df.columns = ['chrom', 'chromStart','chromEnd', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
    "#     tempdf.reset_index(drop=True, inplace=True)\n",
    "#     for index, row in tempdf.iterrows():\n",
    "#         arr.append(file_df[file_df['name'] == row['ID']])\n",
    "#         found_row = file_df[file_df['name'] == row['ID']]\n",
    "#         # print(row['ID'])\n",
    "#         df.loc[df['ID'] == row['ID'], 'chr'] = found_row['chrom'].values[0].replace('chr', '')\n",
    "#         df.loc[df['ID'] == row['ID'], 'start'] = found_row['chromStart'].values[0]\n",
    "#         df.loc[df['ID'] == row['ID'], 'end'] = found_row['chromEnd'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# arr = []\n",
    "# df['chr'] = ''\n",
    "# df['start'] = 0\n",
    "# df['end'] = 0\n",
    "# df['ref'] = 0\n",
    "# df['alt'] = 0\n",
    "# df['region_type'] = 0\n",
    "# df['found'] = 0\n",
    "# file_df = pd.read_csv('data/PCAWG_mutations_to_elements.icgc.public.txt', sep='\\t')\n",
    "# for index, row in df.iterrows():\n",
    "#     found_row = file_df[file_df['reg_id'] == row['ID']]\n",
    "#     if len(found_row) > 1:\n",
    "#         print(\"FOUND\", len(found_row))\n",
    "#     for i, r in found_row.iterrows():\n",
    "#         print(r[['mut_chr', 'mut_pos1', 'mut_pos2', 'mut_ref', 'mut_alt', 'region_type']])\n",
    "#     if len(found_row) > 0:\n",
    "#         df.loc[df['ID'] == row['ID'], 'chr'] = found_row['mut_chr'].values[0].replace('chr', '')\n",
    "#         df.loc[df['ID'] == row['ID'], 'start'] = found_row['mut_pos1'].values[0]\n",
    "#         df.loc[df['ID'] == row['ID'], 'end'] = found_row['mut_pos2'].values[0]\n",
    "#         df.loc[df['ID'] == row['ID'], 'ref'] = found_row['mut_ref'].values[0]\n",
    "#         df.loc[df['ID'] == row['ID'], 'alt'] = found_row['mut_alt'].values[0]\n",
    "#         df.loc[df['ID'] == row['ID'], 'region_type'] = found_row['region_type'].values[0]\n",
    "#         df.loc[df['ID'] == row['ID'], 'found'] = len(found_row)\n",
    "#     break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_data = pd.read_csv('data/dataset_uncensored.csv')\n",
    "raw_data['start'] = raw_data['start'].astype(int)\n",
    "raw_data['end'] = raw_data['end'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlaps = check_overlaps(df, raw_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlaps['target'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "arr_ids = []\n",
    "arr_ids = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/indel'))\n",
    "arr_ids  = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/snv_mnv'))\n",
    "# arr_ids = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/graylist'))\n",
    "arr_ids = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_tcga.controlled/final_consensus_12oct/tcga/indel'))\n",
    "arr_ids  = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_tcga.controlled/final_consensus_12oct/tcga/snv_mnv'))\n",
    "# arr_ids = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_tcga.controlled/final_consensus_12oct/tcga/graylist'))\n",
    "arr_ids = [s for s in arr_ids if '.tbi' not in s]\n",
    "arr_ids_names = [s[:s.find(\".\")] for s in arr_ids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf_gzip(filename):\n",
    "    with gzip.open(filename, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith(b'##')]\n",
    "    df = pd.read_csv(\n",
    "        io.StringIO(b''.join(lines).decode()),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "                'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})\n",
    "    return df\n",
    "    \n",
    "read_vcf_gzip('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/indel/0009b464-b376-4fbc-8a56-da538269a02f.consensus.20161006.somatic.indel.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_files = [val for sublist in [[os.path.join(i[0], j) for j in i[2]] for i in os.walk('D:\\Masters Thesis\\GitHub Repo\\data\\gene-coding-and-non-coding-elements')] for val in sublist]\n",
    "all_files = [s for s in all_files if 'allCombined' not in s]\n",
    "all_files = [s for s in all_files if 'mapFiles' not in s]\n",
    "all_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "all_df = pd.DataFrame()\n",
    "all_df['element'] = ''\n",
    "for file in all_files:\n",
    "    temp_df = pd.read_csv(file, sep='\\t', header=None)\n",
    "    result = re.search('elements\\\\\\\\(.*).bed', file)\n",
    "    temp_df['element'] = str(result.group(1))\n",
    "    all_df = pd.concat([all_df, temp_df])\n",
    "all_df.columns = ['element', 'chrom', 'chromStart','chromEnd', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
    "all_df['chromStart'] = all_df['chromStart'].astype(int)\n",
    "all_df['chromEnd'] = all_df['chromEnd'].astype(int)\n",
    "all_df['score'] = all_df['score'].astype(int)\n",
    "all_df['chrom'] = all_df['chrom'].apply(lambda x: x.replace('chr', ''))\n",
    "\n",
    "all_df.rename({'chrom': 'chr', 'chromStart': 'start', 'chromEnd': 'end'}, inplace=True, axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlaps = check_overlaps_and_annotate(raw_data, all_df, 'element')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlaps[~overlaps['annotation'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlaps1 = check_overlaps(all_df, raw_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlaps1['element'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def row_count(input):\n",
    "    with open(input) as f:\n",
    "        for i, l in enumerate(f):\n",
    "            pass\n",
    "    return i\n",
    "\n",
    "row_count('data/Rheinbay 2020/final_consensus_passonly.snv_mnv_indel.icgc.controlled.maf')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
