{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import gzip\n",
    "import io\n",
    "from scripts.check_overlaps import check_overlaps, check_overlaps_and_annotate\n",
    "from make_dataset import make_dataset_censored,  make_dataset_uncensored, repeat_masker, COSMIC_CGC_interactions, TF_binding_site_annotations, create_vep_input, read_vcf, long_range_interactions_results, create_vep_input, read_vcf, clean_and_preprocess"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rheinbay 2020 Supplementary Table 5<br>\n",
    "We filter on the basis of 'Final filter judgement' = TRUE<br>\n",
    "Then we take the unique element IDs<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel('data/Rheinbay 2020/41586_2020_1965_MOESM4_ESM.xlsx', engine='openpyxl', sheet_name='Table 5 Non-coding point mut dr')\n",
    "# df1 = pd.read_excel('data/Rheinbay 2020/41586_2020_1965_MOESM4_ESM.xlsx', engine='openpyxl', sheet_name='Table 4 Protein-coding point mu')\n",
    "df['file_name'] = df['ID'].apply(lambda x: x.split('::')[0]+'.bed')\n",
    "df['file_name'] = df['file_name'].apply(lambda x: x.replace('mirna_', 'mirna.') if 'mirna_pre' in x or 'mirna_mat' in x else x)\n",
    "noncoding_elements = list(df[df['Final filter judgement'] == True]['ID'].unique())\n",
    "file_df = pd.read_csv('data/PCAWG_mutations_to_elements.icgc.public.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>region_type</th>\n",
       "      <th>reg_id</th>\n",
       "      <th>data_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>111231125</td>\n",
       "      <td>111231125</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>111232885</td>\n",
       "      <td>111232885</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>86866372</td>\n",
       "      <td>86866372</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr7:86865600-86866400::NA::NA</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2267139</td>\n",
       "      <td>2267139</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr1:2267000-2269600::NA::NA</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>111226848</td>\n",
       "      <td>111226848</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901</th>\n",
       "      <td>chr11</td>\n",
       "      <td>65194294</td>\n",
       "      <td>65194295</td>\n",
       "      <td>-</td>\n",
       "      <td>CATAATATTTTAAAA</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "      <td>lncrna.ncrna::gencode::NEAT1::ENSG00000245532.4</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>902</th>\n",
       "      <td>chr11</td>\n",
       "      <td>65269272</td>\n",
       "      <td>65269273</td>\n",
       "      <td>AA</td>\n",
       "      <td>-</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "      <td>lncrna.ncrna::gencode::MALAT1::ENSG00000251562.3</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903</th>\n",
       "      <td>chr11</td>\n",
       "      <td>65267659</td>\n",
       "      <td>65267659</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "      <td>lncrna.ncrna::gencode::MALAT1::ENSG00000251562.3</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904</th>\n",
       "      <td>chr11</td>\n",
       "      <td>65209144</td>\n",
       "      <td>65209148</td>\n",
       "      <td>TTTAT</td>\n",
       "      <td>-</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "      <td>lncrna.ncrna::gencode::NEAT1::ENSG00000245532.4</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>chr11</td>\n",
       "      <td>65196438</td>\n",
       "      <td>65196454</td>\n",
       "      <td>ACATTTTTAAACAAAAA</td>\n",
       "      <td>-</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "      <td>lncrna.ncrna::gencode::NEAT1::ENSG00000245532.4</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>906 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr      start        end                ref              alt  \\\n",
       "0    chr11  111231125  111231125                  G                C   \n",
       "1    chr11  111232885  111232885                  C                T   \n",
       "2     chr7   86866372   86866372                  T                C   \n",
       "3     chr1    2267139    2267139                  G                A   \n",
       "4    chr11  111226848  111226848                  T                C   \n",
       "..     ...        ...        ...                ...              ...   \n",
       "901  chr11   65194294   65194295                  -  CATAATATTTTAAAA   \n",
       "902  chr11   65269272   65269273                 AA                -   \n",
       "903  chr11   65267659   65267659                  A                -   \n",
       "904  chr11   65209144   65209148              TTTAT                -   \n",
       "905  chr11   65196438   65196454  ACATTTTTAAACAAAAA                -   \n",
       "\n",
       "          region_type                                            reg_id  \\\n",
       "0       enhancers.bed      enhancers::chr11:111222000-111233000::NA::NA   \n",
       "1       enhancers.bed      enhancers::chr11:111222000-111233000::NA::NA   \n",
       "2       enhancers.bed         enhancers::chr7:86865600-86866400::NA::NA   \n",
       "3       enhancers.bed           enhancers::chr1:2267000-2269600::NA::NA   \n",
       "4       enhancers.bed      enhancers::chr11:111222000-111233000::NA::NA   \n",
       "..                ...                                               ...   \n",
       "901  lncrna.ncrna.bed   lncrna.ncrna::gencode::NEAT1::ENSG00000245532.4   \n",
       "902  lncrna.ncrna.bed  lncrna.ncrna::gencode::MALAT1::ENSG00000251562.3   \n",
       "903  lncrna.ncrna.bed  lncrna.ncrna::gencode::MALAT1::ENSG00000251562.3   \n",
       "904  lncrna.ncrna.bed   lncrna.ncrna::gencode::NEAT1::ENSG00000245532.4   \n",
       "905  lncrna.ncrna.bed   lncrna.ncrna::gencode::NEAT1::ENSG00000245532.4   \n",
       "\n",
       "             data_source  \n",
       "0    Rheinbay et al 2020  \n",
       "1    Rheinbay et al 2020  \n",
       "2    Rheinbay et al 2020  \n",
       "3    Rheinbay et al 2020  \n",
       "4    Rheinbay et al 2020  \n",
       "..                   ...  \n",
       "901  Rheinbay et al 2020  \n",
       "902  Rheinbay et al 2020  \n",
       "903  Rheinbay et al 2020  \n",
       "904  Rheinbay et al 2020  \n",
       "905  Rheinbay et al 2020  \n",
       "\n",
       "[906 rows x 8 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rheinbay_noncoding_drivers = file_df[file_df['reg_id'].isin(noncoding_elements)][['mut_chr', 'mut_pos1', 'mut_pos2', 'mut_ref', 'mut_alt', 'region_type', 'reg_id']].drop_duplicates().reset_index(drop=True)\n",
    "rheinbay_noncoding_drivers.rename({'mut_chr': 'chr', 'mut_pos1': 'start', 'mut_pos2': 'end', 'mut_ref': 'ref', 'mut_alt': 'alt'}, axis=1, inplace=True)\n",
    "rheinbay_noncoding_drivers['data_source'] = 'Rheinbay et al 2020'\n",
    "rheinbay_noncoding_drivers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>region_type</th>\n",
       "      <th>reg_id</th>\n",
       "      <th>data_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19578253</td>\n",
       "      <td>19578254</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MRTO4</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>19578254</td>\n",
       "      <td>19578255</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MRTO4</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>23889291</td>\n",
       "      <td>23889292</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Ovary</td>\n",
       "      <td>ID3</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>23895459</td>\n",
       "      <td>23895460</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Ovary</td>\n",
       "      <td>ID3</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>23888768</td>\n",
       "      <td>23888769</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Ovary</td>\n",
       "      <td>ID3</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>chr9</td>\n",
       "      <td>125026995</td>\n",
       "      <td>125026996</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Breast</td>\n",
       "      <td>MRRF</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>chr9</td>\n",
       "      <td>124049461</td>\n",
       "      <td>124049462</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Breast</td>\n",
       "      <td>MRRF</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>chrX</td>\n",
       "      <td>70338403</td>\n",
       "      <td>70338404</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>chrX</td>\n",
       "      <td>70401837</td>\n",
       "      <td>70401838</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>chrX</td>\n",
       "      <td>70365659</td>\n",
       "      <td>70365660</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>215 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr      start        end ref alt region_type reg_id  data_source\n",
       "0    chr1   19578253   19578254   G   A       Brain  MRTO4  Dr.Nod 2023\n",
       "1    chr1   19578254   19578255   G   A       Brain  MRTO4  Dr.Nod 2023\n",
       "2    chr1   23889291   23889292   G   T       Ovary    ID3  Dr.Nod 2023\n",
       "3    chr1   23895459   23895460   G   T       Ovary    ID3  Dr.Nod 2023\n",
       "4    chr1   23888768   23888769   T   C       Ovary    ID3  Dr.Nod 2023\n",
       "..    ...        ...        ...  ..  ..         ...    ...          ...\n",
       "210  chr9  125026995  125026996   C   G      Breast   MRRF  Dr.Nod 2023\n",
       "211  chr9  124049461  124049462   T   A      Breast   MRRF  Dr.Nod 2023\n",
       "212  chrX   70338403   70338404   G   A       Brain  MED12  Dr.Nod 2023\n",
       "213  chrX   70401837   70401838   G   T       Brain  MED12  Dr.Nod 2023\n",
       "214  chrX   70365659   70365660   G   A       Brain  MED12  Dr.Nod 2023\n",
       "\n",
       "[215 rows x 8 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dr_nod = pd.read_excel('data\\Dr. Nod\\gkac1251_supplemental_files\\SupplementaryTable4.xlsx', engine='openpyxl')\n",
    "dr_nod = dr_nod[['chr', 'pos0', 'pos1', 'ref', 'alt', 'tissue', 'geneSymbol']]\n",
    "dr_nod.rename({'pos0': 'start', 'pos1': 'end', 'tissue': 'region_type', 'geneSymbol': 'reg_id'}, inplace=True, axis=1)\n",
    "dr_nod['data_source'] = 'Dr.Nod 2023'\n",
    "dr_nod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>region_type</th>\n",
       "      <th>reg_id</th>\n",
       "      <th>data_source</th>\n",
       "      <th>id</th>\n",
       "      <th>driver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>111231125</td>\n",
       "      <td>111231125</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "      <td>test0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11</td>\n",
       "      <td>111232885</td>\n",
       "      <td>111232885</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "      <td>test1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>86866372</td>\n",
       "      <td>86866372</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr7:86865600-86866400::NA::NA</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "      <td>test2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2267139</td>\n",
       "      <td>2267139</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr1:2267000-2269600::NA::NA</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "      <td>test3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11</td>\n",
       "      <td>111226848</td>\n",
       "      <td>111226848</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "      <td>enhancers::chr11:111222000-111233000::NA::NA</td>\n",
       "      <td>Rheinbay et al 2020</td>\n",
       "      <td>test4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1116</th>\n",
       "      <td>9</td>\n",
       "      <td>125026995</td>\n",
       "      <td>125026996</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Breast</td>\n",
       "      <td>MRRF</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "      <td>test1116</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1117</th>\n",
       "      <td>9</td>\n",
       "      <td>124049461</td>\n",
       "      <td>124049462</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Breast</td>\n",
       "      <td>MRRF</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "      <td>test1117</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1118</th>\n",
       "      <td>X</td>\n",
       "      <td>70338403</td>\n",
       "      <td>70338404</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "      <td>test1118</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1119</th>\n",
       "      <td>X</td>\n",
       "      <td>70401837</td>\n",
       "      <td>70401838</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "      <td>test1119</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1120</th>\n",
       "      <td>X</td>\n",
       "      <td>70365659</td>\n",
       "      <td>70365660</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Brain</td>\n",
       "      <td>MED12</td>\n",
       "      <td>Dr.Nod 2023</td>\n",
       "      <td>test1120</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1062 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr      start        end ref alt    region_type  \\\n",
       "0     11  111231125  111231125   G   C  enhancers.bed   \n",
       "1     11  111232885  111232885   C   T  enhancers.bed   \n",
       "2      7   86866372   86866372   T   C  enhancers.bed   \n",
       "3      1    2267139    2267139   G   A  enhancers.bed   \n",
       "4     11  111226848  111226848   T   C  enhancers.bed   \n",
       "...   ..        ...        ...  ..  ..            ...   \n",
       "1116   9  125026995  125026996   C   G         Breast   \n",
       "1117   9  124049461  124049462   T   A         Breast   \n",
       "1118   X   70338403   70338404   G   A          Brain   \n",
       "1119   X   70401837   70401838   G   T          Brain   \n",
       "1120   X   70365659   70365660   G   A          Brain   \n",
       "\n",
       "                                            reg_id          data_source  \\\n",
       "0     enhancers::chr11:111222000-111233000::NA::NA  Rheinbay et al 2020   \n",
       "1     enhancers::chr11:111222000-111233000::NA::NA  Rheinbay et al 2020   \n",
       "2        enhancers::chr7:86865600-86866400::NA::NA  Rheinbay et al 2020   \n",
       "3          enhancers::chr1:2267000-2269600::NA::NA  Rheinbay et al 2020   \n",
       "4     enhancers::chr11:111222000-111233000::NA::NA  Rheinbay et al 2020   \n",
       "...                                            ...                  ...   \n",
       "1116                                          MRRF          Dr.Nod 2023   \n",
       "1117                                          MRRF          Dr.Nod 2023   \n",
       "1118                                         MED12          Dr.Nod 2023   \n",
       "1119                                         MED12          Dr.Nod 2023   \n",
       "1120                                         MED12          Dr.Nod 2023   \n",
       "\n",
       "            id  driver  \n",
       "0        test0       1  \n",
       "1        test1       1  \n",
       "2        test2       1  \n",
       "3        test3       1  \n",
       "4        test4       1  \n",
       "...        ...     ...  \n",
       "1116  test1116       1  \n",
       "1117  test1117       1  \n",
       "1118  test1118       1  \n",
       "1119  test1119       1  \n",
       "1120  test1120       1  \n",
       "\n",
       "[1062 rows x 10 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data = pd.concat([rheinbay_noncoding_drivers, dr_nod])#[(rheinbay_noncoding_drivers['region_type'] == 'gc19_pc.promCore.bed') & (rheinbay_noncoding_drivers['reg_id'].str.contains('TERT'))]#.value_counts()\n",
    "test_data.reset_index(drop=True, inplace=True)\n",
    "test_data['id'] = 'test' + test_data.index.astype(str)\n",
    "test_data['driver'] = 1\n",
    "test_data.drop_duplicates(subset=['chr', 'start', 'end', 'ref', 'alt'], inplace=True)\n",
    "test_data['chr'] = test_data['chr'].apply(lambda x: x.replace('chr', ''))\n",
    "test_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_data.to_csv('./data/test_data_final.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>allele</th>\n",
       "      <th>driver</th>\n",
       "      <th>chrom</th>\n",
       "      <th>itemsReturned</th>\n",
       "      <th>bin</th>\n",
       "      <th>swScore</th>\n",
       "      <th>milliDiv</th>\n",
       "      <th>...</th>\n",
       "      <th>genoEnd</th>\n",
       "      <th>genoLeft</th>\n",
       "      <th>strand</th>\n",
       "      <th>repName</th>\n",
       "      <th>repClass</th>\n",
       "      <th>repFamily</th>\n",
       "      <th>repStart</th>\n",
       "      <th>repEnd</th>\n",
       "      <th>repLeft</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>1295203</td>\n",
       "      <td>1295203</td>\n",
       "      <td>G/A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>594.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1295316.0</td>\n",
       "      <td>-179619944.0</td>\n",
       "      <td>+</td>\n",
       "      <td>(CGGGG)n</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>2.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>1295226</td>\n",
       "      <td>1295226</td>\n",
       "      <td>G/A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>594.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1295316.0</td>\n",
       "      <td>-179619944.0</td>\n",
       "      <td>+</td>\n",
       "      <td>(CGGGG)n</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>2.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>1295226</td>\n",
       "      <td>1295226</td>\n",
       "      <td>G/A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>594.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1295316.0</td>\n",
       "      <td>-179619944.0</td>\n",
       "      <td>+</td>\n",
       "      <td>(CGGGG)n</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>2.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>1295226</td>\n",
       "      <td>1295226</td>\n",
       "      <td>G/A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>594.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1295316.0</td>\n",
       "      <td>-179619944.0</td>\n",
       "      <td>+</td>\n",
       "      <td>(CGGGG)n</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>2.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1295226</td>\n",
       "      <td>1295226</td>\n",
       "      <td>G/A</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>594.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1295316.0</td>\n",
       "      <td>-179619944.0</td>\n",
       "      <td>+</td>\n",
       "      <td>(CGGGG)n</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>2.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>X</td>\n",
       "      <td>72254391</td>\n",
       "      <td>72254391</td>\n",
       "      <td>C/T</td>\n",
       "      <td>0</td>\n",
       "      <td>X</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1136.0</td>\n",
       "      <td>27572.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>...</td>\n",
       "      <td>72257675.0</td>\n",
       "      <td>-83012885.0</td>\n",
       "      <td>-</td>\n",
       "      <td>L1PA3</td>\n",
       "      <td>LINE</td>\n",
       "      <td>L1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6155.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>3</td>\n",
       "      <td>180585989</td>\n",
       "      <td>180585989</td>\n",
       "      <td>C/T</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1962.0</td>\n",
       "      <td>1597.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>...</td>\n",
       "      <td>180586453.0</td>\n",
       "      <td>-17435977.0</td>\n",
       "      <td>-</td>\n",
       "      <td>LTR1C</td>\n",
       "      <td>LTR</td>\n",
       "      <td>ERV1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>656.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>19</td>\n",
       "      <td>30558759</td>\n",
       "      <td>30558759</td>\n",
       "      <td>T/C</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>1.0</td>\n",
       "      <td>818.0</td>\n",
       "      <td>1071.0</td>\n",
       "      <td>167.0</td>\n",
       "      <td>...</td>\n",
       "      <td>30559427.0</td>\n",
       "      <td>-28569556.0</td>\n",
       "      <td>+</td>\n",
       "      <td>(TCCA)n</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>1.0</td>\n",
       "      <td>706.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>16</td>\n",
       "      <td>62818734</td>\n",
       "      <td>62818734</td>\n",
       "      <td>C/T</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1064.0</td>\n",
       "      <td>9249.0</td>\n",
       "      <td>235.0</td>\n",
       "      <td>...</td>\n",
       "      <td>62822231.0</td>\n",
       "      <td>-27532522.0</td>\n",
       "      <td>-</td>\n",
       "      <td>L1MB8</td>\n",
       "      <td>LINE</td>\n",
       "      <td>L1</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>7037.0</td>\n",
       "      <td>3103.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>19</td>\n",
       "      <td>3299586</td>\n",
       "      <td>3299586</td>\n",
       "      <td>T/C</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>1.0</td>\n",
       "      <td>610.0</td>\n",
       "      <td>1204.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3299670.0</td>\n",
       "      <td>-55829313.0</td>\n",
       "      <td>+</td>\n",
       "      <td>(TTC)n</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>1.0</td>\n",
       "      <td>180.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>217 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr      start        end allele  driver chrom  itemsReturned     bin  \\\n",
       "0     5    1295203    1295203    G/A       1     5            1.0   594.0   \n",
       "1     5    1295226    1295226    G/A       1     5            1.0   594.0   \n",
       "2     5    1295226    1295226    G/A       1     5            1.0   594.0   \n",
       "3     5    1295226    1295226    G/A       1     5            1.0   594.0   \n",
       "4     5    1295226    1295226    G/A       1     5            1.0   594.0   \n",
       "..   ..        ...        ...    ...     ...   ...            ...     ...   \n",
       "212   X   72254391   72254391    C/T       0     X            1.0  1136.0   \n",
       "213   3  180585989  180585989    C/T       0     3            1.0  1962.0   \n",
       "214  19   30558759   30558759    T/C       0    19            1.0   818.0   \n",
       "215  16   62818734   62818734    C/T       0    16            1.0  1064.0   \n",
       "216  19    3299586    3299586    T/C       0    19            1.0   610.0   \n",
       "\n",
       "     swScore  milliDiv  ...      genoEnd     genoLeft strand   repName  \\\n",
       "0      289.0     254.0  ...    1295316.0 -179619944.0      +  (CGGGG)n   \n",
       "1      289.0     254.0  ...    1295316.0 -179619944.0      +  (CGGGG)n   \n",
       "2      289.0     254.0  ...    1295316.0 -179619944.0      +  (CGGGG)n   \n",
       "3      289.0     254.0  ...    1295316.0 -179619944.0      +  (CGGGG)n   \n",
       "4      289.0     254.0  ...    1295316.0 -179619944.0      +  (CGGGG)n   \n",
       "..       ...       ...  ...          ...          ...    ...       ...   \n",
       "212  27572.0      45.0  ...   72257675.0  -83012885.0      -     L1PA3   \n",
       "213   1597.0     174.0  ...  180586453.0  -17435977.0      -     LTR1C   \n",
       "214   1071.0     167.0  ...   30559427.0  -28569556.0      +   (TCCA)n   \n",
       "215   9249.0     235.0  ...   62822231.0  -27532522.0      -     L1MB8   \n",
       "216   1204.0      89.0  ...    3299670.0  -55829313.0      +    (TTC)n   \n",
       "\n",
       "          repClass      repFamily repStart  repEnd repLeft id  \n",
       "0    Simple_repeat  Simple_repeat      2.0   133.0     0.0  2  \n",
       "1    Simple_repeat  Simple_repeat      2.0   133.0     0.0  2  \n",
       "2    Simple_repeat  Simple_repeat      2.0   133.0     0.0  2  \n",
       "3    Simple_repeat  Simple_repeat      2.0   133.0     0.0  2  \n",
       "4    Simple_repeat  Simple_repeat      2.0   133.0     0.0  2  \n",
       "..             ...            ...      ...     ...     ... ..  \n",
       "212           LINE             L1      0.0  6155.0     1.0  4  \n",
       "213            LTR           ERV1      0.0   656.0     1.0  2  \n",
       "214  Simple_repeat  Simple_repeat      1.0   706.0     0.0  1  \n",
       "215           LINE             L1     -1.0  7037.0  3103.0  1  \n",
       "216  Simple_repeat  Simple_repeat      1.0   180.0     0.0  1  \n",
       "\n",
       "[217 rows x 24 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_pickle('data/repeat_masker.pickle')\n",
    "# df = repeat_masker(df)\n",
    "# df = COSMIC_CGC_interactions(df)\n",
    "# df = TF_binding_site_annotations(df)\n",
    "# df = long_range_interactions_results(df)\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "arr_ids = []\n",
    "arr_ids = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/indel'))\n",
    "arr_ids  = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/snv_mnv'))\n",
    "# arr_ids = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/graylist'))\n",
    "arr_ids = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_tcga.controlled/final_consensus_12oct/tcga/indel'))\n",
    "arr_ids  = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_tcga.controlled/final_consensus_12oct/tcga/snv_mnv'))\n",
    "# arr_ids = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_tcga.controlled/final_consensus_12oct/tcga/graylist'))\n",
    "arr_ids = [s for s in arr_ids if '.tbi' not in s]\n",
    "arr_ids_names = [s[:s.find(\".\")] for s in arr_ids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf_gzip(filename):\n",
    "    with gzip.open(filename, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith(b'##')]\n",
    "    df = pd.read_csv(\n",
    "        io.StringIO(b''.join(lines).decode()),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "                'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})\n",
    "    return df\n",
    "    \n",
    "read_vcf_gzip('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/indel/0009b464-b376-4fbc-8a56-da538269a02f.consensus.20161006.somatic.indel.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23159591"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def row_count(input):\n",
    "    with open(input) as f:\n",
    "        for i, l in enumerate(f):\n",
    "            pass\n",
    "    return i\n",
    "\n",
    "row_count('data/Rheinbay 2020/final_consensus_passonly.snv_mnv_indel.icgc.controlled.maf')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
