{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import gzip\n",
    "import io\n",
    "from scripts.check_overlaps import check_overlaps, check_overlaps_and_annotate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rheinbay 2020 Supplementary Table 5<br>\n",
    "We filter on the basis of 'Final filter judgement' = TRUE<br>\n",
    "Then we take the unique element IDs<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel('data/Rheinbay 2020/41586_2020_1965_MOESM4_ESM.xlsx', engine='openpyxl', sheet_name='Table 5 Non-coding point mut dr')\n",
    "# df1 = pd.read_excel('data/Rheinbay 2020/41586_2020_1965_MOESM4_ESM.xlsx', engine='openpyxl', sheet_name='Table 4 Protein-coding point mu')\n",
    "df['file_name'] = df['ID'].apply(lambda x: x.split('::')[0]+'.bed')\n",
    "df['file_name'] = df['file_name'].apply(lambda x: x.replace('mirna_', 'mirna.') if 'mirna_pre' in x or 'mirna_mat' in x else x)\n",
    "noncoding_elements = list(df[df['Final filter judgement'] == True]['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_df = pd.read_csv('data/PCAWG_mutations_to_elements.icgc.public.txt', sep='\\t')\n",
    "file_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>region_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>111231125</td>\n",
       "      <td>111231125</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>111232885</td>\n",
       "      <td>111232885</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>enhancers.bed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>86866372</td>\n",
       "      <td>86866372</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2267139</td>\n",
       "      <td>2267139</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>enhancers.bed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>111226848</td>\n",
       "      <td>111226848</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>enhancers.bed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>895</th>\n",
       "      <td>chr11</td>\n",
       "      <td>65194294</td>\n",
       "      <td>65194295</td>\n",
       "      <td>-</td>\n",
       "      <td>CATAATATTTTAAAA</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>896</th>\n",
       "      <td>chr11</td>\n",
       "      <td>65269272</td>\n",
       "      <td>65269273</td>\n",
       "      <td>AA</td>\n",
       "      <td>-</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>897</th>\n",
       "      <td>chr11</td>\n",
       "      <td>65267659</td>\n",
       "      <td>65267659</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>898</th>\n",
       "      <td>chr11</td>\n",
       "      <td>65209144</td>\n",
       "      <td>65209148</td>\n",
       "      <td>TTTAT</td>\n",
       "      <td>-</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899</th>\n",
       "      <td>chr11</td>\n",
       "      <td>65196438</td>\n",
       "      <td>65196454</td>\n",
       "      <td>ACATTTTTAAACAAAAA</td>\n",
       "      <td>-</td>\n",
       "      <td>lncrna.ncrna.bed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>900 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr      start        end                ref              alt  \\\n",
       "0    chr11  111231125  111231125                  G                C   \n",
       "1    chr11  111232885  111232885                  C                T   \n",
       "2     chr7   86866372   86866372                  T                C   \n",
       "3     chr1    2267139    2267139                  G                A   \n",
       "4    chr11  111226848  111226848                  T                C   \n",
       "..     ...        ...        ...                ...              ...   \n",
       "895  chr11   65194294   65194295                  -  CATAATATTTTAAAA   \n",
       "896  chr11   65269272   65269273                 AA                -   \n",
       "897  chr11   65267659   65267659                  A                -   \n",
       "898  chr11   65209144   65209148              TTTAT                -   \n",
       "899  chr11   65196438   65196454  ACATTTTTAAACAAAAA                -   \n",
       "\n",
       "          region_type  \n",
       "0       enhancers.bed  \n",
       "1       enhancers.bed  \n",
       "2       enhancers.bed  \n",
       "3       enhancers.bed  \n",
       "4       enhancers.bed  \n",
       "..                ...  \n",
       "895  lncrna.ncrna.bed  \n",
       "896  lncrna.ncrna.bed  \n",
       "897  lncrna.ncrna.bed  \n",
       "898  lncrna.ncrna.bed  \n",
       "899  lncrna.ncrna.bed  \n",
       "\n",
       "[900 rows x 6 columns]"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rheinbay_noncoding_drivers = file_df[file_df['reg_id'].isin(noncoding_elements)][['mut_chr', 'mut_pos1', 'mut_pos2', 'mut_ref', 'mut_alt', 'region_type']].drop_duplicates().reset_index(drop=True)\n",
    "rheinbay_noncoding_drivers.rename({'mut_chr': 'chr', 'mut_pos1': 'start', 'mut_pos2': 'end', 'mut_ref': 'ref', 'mut_alt': 'alt'}, axis=1, inplace=True)\n",
    "rheinbay_noncoding_drivers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "elements_list = os.listdir('data/gene-coding-and-non-coding-elements/')\n",
    "elements_list.remove('allCombined.zip')\n",
    "elements_list.remove('mapFiles')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "enhancers.bed\n",
      "gc19_pc.3utr.bed\n",
      "gc19_pc.5utr.bed\n",
      "gc19_pc.cds.bed\n",
      "gc19_pc.promCore.bed\n",
      "gc19_pc.promDomain.bed\n",
      "gc19_pc.ss.bed\n",
      "lncrna.ncrna.bed\n",
      "lncrna.promCore.bed\n",
      "lncrna.promDomain.bed\n",
      "lncrna.ss.bed\n",
      "mirna.mat.bed\n",
      "mirna.pre.bed\n",
      "mirna.prom.bed\n",
      "smallrna.ncrna.bed\n",
      "supplementary_lncrna.ncrna.bed\n",
      "supplementary_lncrna.prom.bed\n",
      "supplementary_lncrna.ss.bed\n"
     ]
    }
   ],
   "source": [
    "arr = []\n",
    "df['chr'] = ''\n",
    "df['start'] = 0\n",
    "df['end'] = 0\n",
    "for element in elements_list:\n",
    "    print(element)\n",
    "    tempdf = df[df['file_name'] == element]\n",
    "    file_df = pd.read_csv('data/gene-coding-and-non-coding-elements/'+element, sep='\\t', header=None)\n",
    "    file_df.columns = ['chrom', 'chromStart','chromEnd', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
    "    tempdf.reset_index(drop=True, inplace=True)\n",
    "    for index, row in tempdf.iterrows():\n",
    "        arr.append(file_df[file_df['name'] == row['ID']])\n",
    "        found_row = file_df[file_df['name'] == row['ID']]\n",
    "        # print(row['ID'])\n",
    "        df.loc[df['ID'] == row['ID'], 'chr'] = found_row['chrom'].values[0].replace('chr', '')\n",
    "        df.loc[df['ID'] == row['ID'], 'start'] = found_row['chromStart'].values[0]\n",
    "        df.loc[df['ID'] == row['ID'], 'end'] = found_row['chromEnd'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region_type</th>\n",
       "      <th>mut_chr</th>\n",
       "      <th>mut_pos1</th>\n",
       "      <th>mut_pos2</th>\n",
       "      <th>mut_ref</th>\n",
       "      <th>mut_alt</th>\n",
       "      <th>mut_patient</th>\n",
       "      <th>mut_cc</th>\n",
       "      <th>mut_tag</th>\n",
       "      <th>reg_chr</th>\n",
       "      <th>reg_starts</th>\n",
       "      <th>reg_ends</th>\n",
       "      <th>reg_id</th>\n",
       "      <th>reg_frag_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>264519</th>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>chr20</td>\n",
       "      <td>32398974</td>\n",
       "      <td>32398974</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>DO49087</td>\n",
       "      <td>Panc-AdenoCA</td>\n",
       "      <td>AAG&gt;G</td>\n",
       "      <td>chr20</td>\n",
       "      <td>32398909.0</td>\n",
       "      <td>32399274.0</td>\n",
       "      <td>gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266069</th>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>chr20</td>\n",
       "      <td>32399061</td>\n",
       "      <td>32399061</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>DO35230</td>\n",
       "      <td>Panc-AdenoCA</td>\n",
       "      <td>GTG&gt;C</td>\n",
       "      <td>chr20</td>\n",
       "      <td>32398909.0</td>\n",
       "      <td>32399274.0</td>\n",
       "      <td>gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266105</th>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>chr20</td>\n",
       "      <td>32399051</td>\n",
       "      <td>32399051</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>DO218442</td>\n",
       "      <td>Stomach-AdenoCA</td>\n",
       "      <td>GTT&gt;G</td>\n",
       "      <td>chr20</td>\n",
       "      <td>32398909.0</td>\n",
       "      <td>32399274.0</td>\n",
       "      <td>gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266373</th>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>chr20</td>\n",
       "      <td>32398974</td>\n",
       "      <td>32398974</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>DO218223</td>\n",
       "      <td>Stomach-AdenoCA</td>\n",
       "      <td>AAG&gt;G</td>\n",
       "      <td>chr20</td>\n",
       "      <td>32398909.0</td>\n",
       "      <td>32399274.0</td>\n",
       "      <td>gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271428</th>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>chr20</td>\n",
       "      <td>32398939</td>\n",
       "      <td>32398939</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>DO50796</td>\n",
       "      <td>Liver-HCC</td>\n",
       "      <td>CAG&gt;T</td>\n",
       "      <td>chr20</td>\n",
       "      <td>32398909.0</td>\n",
       "      <td>32399274.0</td>\n",
       "      <td>gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 region_type mut_chr  mut_pos1  mut_pos2 mut_ref mut_alt  \\\n",
       "264519  gc19_pc.promCore.bed   chr20  32398974  32398974       A       G   \n",
       "266069  gc19_pc.promCore.bed   chr20  32399061  32399061       T       C   \n",
       "266105  gc19_pc.promCore.bed   chr20  32399051  32399051       T       G   \n",
       "266373  gc19_pc.promCore.bed   chr20  32398974  32398974       A       G   \n",
       "271428  gc19_pc.promCore.bed   chr20  32398939  32398939       A       T   \n",
       "\n",
       "       mut_patient           mut_cc mut_tag reg_chr  reg_starts    reg_ends  \\\n",
       "264519     DO49087     Panc-AdenoCA   AAG>G   chr20  32398909.0  32399274.0   \n",
       "266069     DO35230     Panc-AdenoCA   GTG>C   chr20  32398909.0  32399274.0   \n",
       "266105    DO218442  Stomach-AdenoCA   GTT>G   chr20  32398909.0  32399274.0   \n",
       "266373    DO218223  Stomach-AdenoCA   AAG>G   chr20  32398909.0  32399274.0   \n",
       "271428     DO50796        Liver-HCC   CAG>T   chr20  32398909.0  32399274.0   \n",
       "\n",
       "                                                   reg_id  reg_frag_id  \n",
       "264519  gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...            1  \n",
       "266069  gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...            1  \n",
       "266105  gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...            1  \n",
       "266373  gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...            1  \n",
       "271428  gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...            1  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_df[file_df['reg_id'] == 'gc19_pc.promCore::gencode::CHMP4B::ENSG00000101421.3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FOUND 5\n",
      "mut_chr                       chr20\n",
      "mut_pos1                   32398974\n",
      "mut_pos2                   32398974\n",
      "mut_ref                           A\n",
      "mut_alt                           G\n",
      "region_type    gc19_pc.promCore.bed\n",
      "Name: 264519, dtype: object\n",
      "mut_chr                       chr20\n",
      "mut_pos1                   32399061\n",
      "mut_pos2                   32399061\n",
      "mut_ref                           T\n",
      "mut_alt                           C\n",
      "region_type    gc19_pc.promCore.bed\n",
      "Name: 266069, dtype: object\n",
      "mut_chr                       chr20\n",
      "mut_pos1                   32399051\n",
      "mut_pos2                   32399051\n",
      "mut_ref                           T\n",
      "mut_alt                           G\n",
      "region_type    gc19_pc.promCore.bed\n",
      "Name: 266105, dtype: object\n",
      "mut_chr                       chr20\n",
      "mut_pos1                   32398974\n",
      "mut_pos2                   32398974\n",
      "mut_ref                           A\n",
      "mut_alt                           G\n",
      "region_type    gc19_pc.promCore.bed\n",
      "Name: 266373, dtype: object\n",
      "mut_chr                       chr20\n",
      "mut_pos1                   32398939\n",
      "mut_pos2                   32398939\n",
      "mut_ref                           A\n",
      "mut_alt                           T\n",
      "region_type    gc19_pc.promCore.bed\n",
      "Name: 271428, dtype: object\n"
     ]
    }
   ],
   "source": [
    "arr = []\n",
    "df['chr'] = ''\n",
    "df['start'] = 0\n",
    "df['end'] = 0\n",
    "df['ref'] = 0\n",
    "df['alt'] = 0\n",
    "df['region_type'] = 0\n",
    "df['found'] = 0\n",
    "file_df = pd.read_csv('data/PCAWG_mutations_to_elements.icgc.public.txt', sep='\\t')\n",
    "for index, row in df.iterrows():\n",
    "    found_row = file_df[file_df['reg_id'] == row['ID']]\n",
    "    if len(found_row) > 1:\n",
    "        print(\"FOUND\", len(found_row))\n",
    "    for i, r in found_row.iterrows():\n",
    "        print(r[['mut_chr', 'mut_pos1', 'mut_pos2', 'mut_ref', 'mut_alt', 'region_type']])\n",
    "    if len(found_row) > 0:\n",
    "        df.loc[df['ID'] == row['ID'], 'chr'] = found_row['mut_chr'].values[0].replace('chr', '')\n",
    "        df.loc[df['ID'] == row['ID'], 'start'] = found_row['mut_pos1'].values[0]\n",
    "        df.loc[df['ID'] == row['ID'], 'end'] = found_row['mut_pos2'].values[0]\n",
    "        df.loc[df['ID'] == row['ID'], 'ref'] = found_row['mut_ref'].values[0]\n",
    "        df.loc[df['ID'] == row['ID'], 'alt'] = found_row['mut_alt'].values[0]\n",
    "        df.loc[df['ID'] == row['ID'], 'region_type'] = found_row['region_type'].values[0]\n",
    "        df.loc[df['ID'] == row['ID'], 'found'] = len(found_row)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>target</th>\n",
       "      <th>tissue</th>\n",
       "      <th>gene</th>\n",
       "      <th>Brown_observed</th>\n",
       "      <th>max_method_contr</th>\n",
       "      <th>max_method_id</th>\n",
       "      <th>Pre-filter q-value</th>\n",
       "      <th>Post-filter q-value</th>\n",
       "      <th>Mutations</th>\n",
       "      <th>...</th>\n",
       "      <th>Mutated patients w. expression data</th>\n",
       "      <th>Wild-type patients with expression data</th>\n",
       "      <th>file_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>region_type</th>\n",
       "      <th>found</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...</td>\n",
       "      <td>promCore</td>\n",
       "      <td>meta_Adenocarcinoma</td>\n",
       "      <td>CHMP4B</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.373270</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.413950</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>765.0</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gc19_pc.promCore::gencode::DYNLL1::ENSG0000008...</td>\n",
       "      <td>promCore</td>\n",
       "      <td>meta_Adenocarcinoma</td>\n",
       "      <td>DYNLL1</td>\n",
       "      <td>0.000248</td>\n",
       "      <td>0.371018</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.489554</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>8.0</td>\n",
       "      <td>761.0</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gc19_pc.promCore::gencode::HES1::ENSG000001143...</td>\n",
       "      <td>promCore</td>\n",
       "      <td>meta_Adenocarcinoma</td>\n",
       "      <td>HES1</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.378151</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.293439</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>763.0</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gc19_pc.promCore::gencode::KB-1507C5.2::ENSG00...</td>\n",
       "      <td>promCore</td>\n",
       "      <td>meta_Adenocarcinoma</td>\n",
       "      <td>KB-1507C5.2</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>0.374191</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.388602</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>767.0</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gc19_pc.promCore::gencode::LEPROTL1::ENSG00000...</td>\n",
       "      <td>promCore</td>\n",
       "      <td>meta_Adenocarcinoma</td>\n",
       "      <td>LEPROTL1</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>0.375440</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.361455</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>764.0</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>smallrna.ncrna::trnascanse::Asn.GTT::id.trna.578</td>\n",
       "      <td>smallrna</td>\n",
       "      <td>Thy-AdenoCa</td>\n",
       "      <td>Asn.GTT</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>0.359477</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.247733</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>smallrna.ncrna.bed</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1204</th>\n",
       "      <td>smallrna.ncrna::trnascanse::Asp.GTC::id.trna.253</td>\n",
       "      <td>smallrna</td>\n",
       "      <td>Thy-AdenoCa</td>\n",
       "      <td>Asp.GTC</td>\n",
       "      <td>0.000238</td>\n",
       "      <td>0.348613</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.442825</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>smallrna.ncrna.bed</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1205</th>\n",
       "      <td>smallrna.ncrna::trnascanse::Cys.GCA::id.trna.394</td>\n",
       "      <td>smallrna</td>\n",
       "      <td>Thy-AdenoCa</td>\n",
       "      <td>Cys.GCA</td>\n",
       "      <td>0.000197</td>\n",
       "      <td>0.350368</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.414456</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>smallrna.ncrna.bed</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1206</th>\n",
       "      <td>smallrna.ncrna::gencode::SNORD115-17::ENSG0000...</td>\n",
       "      <td>smallrna</td>\n",
       "      <td>Uterus-AdenoCa</td>\n",
       "      <td>SNORD115-17</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.380399</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.246445</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>smallrna.ncrna.bed</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207</th>\n",
       "      <td>smallrna.ncrna::rfam::tRNA::RF00005.27.8659</td>\n",
       "      <td>smallrna</td>\n",
       "      <td>Uterus-AdenoCa</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.386719</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.244590</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>smallrna.ncrna.bed</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1208 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     ID    target  \\\n",
       "0     gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...  promCore   \n",
       "1     gc19_pc.promCore::gencode::DYNLL1::ENSG0000008...  promCore   \n",
       "2     gc19_pc.promCore::gencode::HES1::ENSG000001143...  promCore   \n",
       "3     gc19_pc.promCore::gencode::KB-1507C5.2::ENSG00...  promCore   \n",
       "4     gc19_pc.promCore::gencode::LEPROTL1::ENSG00000...  promCore   \n",
       "...                                                 ...       ...   \n",
       "1203   smallrna.ncrna::trnascanse::Asn.GTT::id.trna.578  smallrna   \n",
       "1204   smallrna.ncrna::trnascanse::Asp.GTC::id.trna.253  smallrna   \n",
       "1205   smallrna.ncrna::trnascanse::Cys.GCA::id.trna.394  smallrna   \n",
       "1206  smallrna.ncrna::gencode::SNORD115-17::ENSG0000...  smallrna   \n",
       "1207        smallrna.ncrna::rfam::tRNA::RF00005.27.8659  smallrna   \n",
       "\n",
       "                   tissue         gene  Brown_observed  max_method_contr  \\\n",
       "0     meta_Adenocarcinoma       CHMP4B        0.000194          0.373270   \n",
       "1     meta_Adenocarcinoma       DYNLL1        0.000248          0.371018   \n",
       "2     meta_Adenocarcinoma         HES1        0.000109          0.378151   \n",
       "3     meta_Adenocarcinoma  KB-1507C5.2        0.000175          0.374191   \n",
       "4     meta_Adenocarcinoma     LEPROTL1        0.000151          0.375440   \n",
       "...                   ...          ...             ...               ...   \n",
       "1203          Thy-AdenoCa      Asn.GTT        0.000068          0.359477   \n",
       "1204          Thy-AdenoCa      Asp.GTC        0.000238          0.348613   \n",
       "1205          Thy-AdenoCa      Cys.GCA        0.000197          0.350368   \n",
       "1206       Uterus-AdenoCa  SNORD115-17        0.000066          0.380399   \n",
       "1207       Uterus-AdenoCa         tRNA        0.000065          0.386719   \n",
       "\n",
       "        max_method_id  Pre-filter q-value  Post-filter q-value  Mutations  \\\n",
       "0     Fisher_observed            0.413950                  1.0          8   \n",
       "1     Fisher_observed            0.489554                  1.0         15   \n",
       "2     Fisher_observed            0.293439                  1.0          9   \n",
       "3     Fisher_observed            0.388602                  1.0          9   \n",
       "4     Fisher_observed            0.361455                  1.0          9   \n",
       "...               ...                 ...                  ...        ...   \n",
       "1203  Fisher_observed            0.247733                  1.0          2   \n",
       "1204  Fisher_observed            0.442825                  1.0          1   \n",
       "1205  Fisher_observed            0.414456                  1.0          1   \n",
       "1206  Fisher_observed            0.246445                  1.0          2   \n",
       "1207  Fisher_observed            0.244590                  1.0          2   \n",
       "\n",
       "      ...  Mutated patients w. expression data  \\\n",
       "0     ...                                  4.0   \n",
       "1     ...                                  8.0   \n",
       "2     ...                                  6.0   \n",
       "3     ...                                  2.0   \n",
       "4     ...                                  5.0   \n",
       "...   ...                                  ...   \n",
       "1203  ...                                  NaN   \n",
       "1204  ...                                  NaN   \n",
       "1205  ...                                  NaN   \n",
       "1206  ...                                  NaN   \n",
       "1207  ...                                  NaN   \n",
       "\n",
       "      Wild-type patients with expression data             file_name   chr  \\\n",
       "0                                       765.0  gc19_pc.promCore.bed  None   \n",
       "1                                       761.0  gc19_pc.promCore.bed  None   \n",
       "2                                       763.0  gc19_pc.promCore.bed  None   \n",
       "3                                       767.0  gc19_pc.promCore.bed  None   \n",
       "4                                       764.0  gc19_pc.promCore.bed  None   \n",
       "...                                       ...                   ...   ...   \n",
       "1203                                      NaN    smallrna.ncrna.bed    []   \n",
       "1204                                      NaN    smallrna.ncrna.bed    []   \n",
       "1205                                      NaN    smallrna.ncrna.bed    []   \n",
       "1206                                      NaN    smallrna.ncrna.bed    []   \n",
       "1207                                      NaN    smallrna.ncrna.bed    []   \n",
       "\n",
       "      start   end   ref   alt  region_type  found  \n",
       "0      None  None  None  None         None      5  \n",
       "1      None  None  None  None         None      9  \n",
       "2      None  None  None  None         None      3  \n",
       "3      None  None  None  None         None      6  \n",
       "4      None  None  None  None         None      4  \n",
       "...     ...   ...   ...   ...          ...    ...  \n",
       "1203     []    []    []    []           []      0  \n",
       "1204     []    []    []    []           []      0  \n",
       "1205     []    []    []    []           []      0  \n",
       "1206     []    []    []    []           []      0  \n",
       "1207     []    []    []    []           []      0  \n",
       "\n",
       "[1208 rows x 33 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df#[df['start']>0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>target</th>\n",
       "      <th>tissue</th>\n",
       "      <th>gene</th>\n",
       "      <th>Brown_observed</th>\n",
       "      <th>max_method_contr</th>\n",
       "      <th>max_method_id</th>\n",
       "      <th>Pre-filter q-value</th>\n",
       "      <th>Post-filter q-value</th>\n",
       "      <th>Mutations</th>\n",
       "      <th>...</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>Final filter judgement</th>\n",
       "      <th>Expression_p-value</th>\n",
       "      <th>Expression_Fold Difference</th>\n",
       "      <th>Mutated patients w. expression data</th>\n",
       "      <th>Wild-type patients with expression data</th>\n",
       "      <th>file_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...</td>\n",
       "      <td>promCore</td>\n",
       "      <td>meta_Adenocarcinoma</td>\n",
       "      <td>CHMP4B</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.373270</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.413950</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.702076</td>\n",
       "      <td>1.216133</td>\n",
       "      <td>4.0</td>\n",
       "      <td>765.0</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>20</td>\n",
       "      <td>32398909</td>\n",
       "      <td>32399274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gc19_pc.promCore::gencode::DYNLL1::ENSG0000008...</td>\n",
       "      <td>promCore</td>\n",
       "      <td>meta_Adenocarcinoma</td>\n",
       "      <td>DYNLL1</td>\n",
       "      <td>0.000248</td>\n",
       "      <td>0.371018</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.489554</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.694475</td>\n",
       "      <td>0.991591</td>\n",
       "      <td>8.0</td>\n",
       "      <td>761.0</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>12</td>\n",
       "      <td>120907452</td>\n",
       "      <td>120934115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gc19_pc.promCore::gencode::HES1::ENSG000001143...</td>\n",
       "      <td>promCore</td>\n",
       "      <td>meta_Adenocarcinoma</td>\n",
       "      <td>HES1</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.378151</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.293439</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0.242469</td>\n",
       "      <td>0.546071</td>\n",
       "      <td>6.0</td>\n",
       "      <td>763.0</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>3</td>\n",
       "      <td>193853733</td>\n",
       "      <td>193854133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gc19_pc.promCore::gencode::KB-1507C5.2::ENSG00...</td>\n",
       "      <td>promCore</td>\n",
       "      <td>meta_Adenocarcinoma</td>\n",
       "      <td>KB-1507C5.2</td>\n",
       "      <td>0.000175</td>\n",
       "      <td>0.374191</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.388602</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0.231362</td>\n",
       "      <td>1.838830</td>\n",
       "      <td>2.0</td>\n",
       "      <td>767.0</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>8</td>\n",
       "      <td>103876327</td>\n",
       "      <td>103876578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gc19_pc.promCore::gencode::LEPROTL1::ENSG00000...</td>\n",
       "      <td>promCore</td>\n",
       "      <td>meta_Adenocarcinoma</td>\n",
       "      <td>LEPROTL1</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>0.375440</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.361455</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.050680</td>\n",
       "      <td>1.203716</td>\n",
       "      <td>5.0</td>\n",
       "      <td>764.0</td>\n",
       "      <td>gc19_pc.promCore.bed</td>\n",
       "      <td>8</td>\n",
       "      <td>29952713</td>\n",
       "      <td>29959655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>smallrna.ncrna::trnascanse::Asn.GTT::id.trna.578</td>\n",
       "      <td>smallrna</td>\n",
       "      <td>Thy-AdenoCa</td>\n",
       "      <td>Asn.GTT</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>0.359477</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.247733</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>smallrna.ncrna.bed</td>\n",
       "      <td>17</td>\n",
       "      <td>36908033</td>\n",
       "      <td>36908107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1204</th>\n",
       "      <td>smallrna.ncrna::trnascanse::Asp.GTC::id.trna.253</td>\n",
       "      <td>smallrna</td>\n",
       "      <td>Thy-AdenoCa</td>\n",
       "      <td>Asp.GTC</td>\n",
       "      <td>0.000238</td>\n",
       "      <td>0.348613</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.442825</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>smallrna.ncrna.bed</td>\n",
       "      <td>6</td>\n",
       "      <td>27447452</td>\n",
       "      <td>27447524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1205</th>\n",
       "      <td>smallrna.ncrna::trnascanse::Cys.GCA::id.trna.394</td>\n",
       "      <td>smallrna</td>\n",
       "      <td>Thy-AdenoCa</td>\n",
       "      <td>Cys.GCA</td>\n",
       "      <td>0.000197</td>\n",
       "      <td>0.350368</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.414456</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>smallrna.ncrna.bed</td>\n",
       "      <td>7</td>\n",
       "      <td>149295045</td>\n",
       "      <td>149295117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1206</th>\n",
       "      <td>smallrna.ncrna::gencode::SNORD115-17::ENSG0000...</td>\n",
       "      <td>smallrna</td>\n",
       "      <td>Uterus-AdenoCa</td>\n",
       "      <td>SNORD115-17</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.380399</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.246445</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>smallrna.ncrna.bed</td>\n",
       "      <td>15</td>\n",
       "      <td>25446469</td>\n",
       "      <td>25446551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207</th>\n",
       "      <td>smallrna.ncrna::rfam::tRNA::RF00005.27.8659</td>\n",
       "      <td>smallrna</td>\n",
       "      <td>Uterus-AdenoCa</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.386719</td>\n",
       "      <td>Fisher_observed</td>\n",
       "      <td>0.244590</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>smallrna.ncrna.bed</td>\n",
       "      <td>X</td>\n",
       "      <td>125864211</td>\n",
       "      <td>125864278</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1208 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     ID    target  \\\n",
       "0     gc19_pc.promCore::gencode::CHMP4B::ENSG0000010...  promCore   \n",
       "1     gc19_pc.promCore::gencode::DYNLL1::ENSG0000008...  promCore   \n",
       "2     gc19_pc.promCore::gencode::HES1::ENSG000001143...  promCore   \n",
       "3     gc19_pc.promCore::gencode::KB-1507C5.2::ENSG00...  promCore   \n",
       "4     gc19_pc.promCore::gencode::LEPROTL1::ENSG00000...  promCore   \n",
       "...                                                 ...       ...   \n",
       "1203   smallrna.ncrna::trnascanse::Asn.GTT::id.trna.578  smallrna   \n",
       "1204   smallrna.ncrna::trnascanse::Asp.GTC::id.trna.253  smallrna   \n",
       "1205   smallrna.ncrna::trnascanse::Cys.GCA::id.trna.394  smallrna   \n",
       "1206  smallrna.ncrna::gencode::SNORD115-17::ENSG0000...  smallrna   \n",
       "1207        smallrna.ncrna::rfam::tRNA::RF00005.27.8659  smallrna   \n",
       "\n",
       "                   tissue         gene  Brown_observed  max_method_contr  \\\n",
       "0     meta_Adenocarcinoma       CHMP4B        0.000194          0.373270   \n",
       "1     meta_Adenocarcinoma       DYNLL1        0.000248          0.371018   \n",
       "2     meta_Adenocarcinoma         HES1        0.000109          0.378151   \n",
       "3     meta_Adenocarcinoma  KB-1507C5.2        0.000175          0.374191   \n",
       "4     meta_Adenocarcinoma     LEPROTL1        0.000151          0.375440   \n",
       "...                   ...          ...             ...               ...   \n",
       "1203          Thy-AdenoCa      Asn.GTT        0.000068          0.359477   \n",
       "1204          Thy-AdenoCa      Asp.GTC        0.000238          0.348613   \n",
       "1205          Thy-AdenoCa      Cys.GCA        0.000197          0.350368   \n",
       "1206       Uterus-AdenoCa  SNORD115-17        0.000066          0.380399   \n",
       "1207       Uterus-AdenoCa         tRNA        0.000065          0.386719   \n",
       "\n",
       "        max_method_id  Pre-filter q-value  Post-filter q-value  Mutations  \\\n",
       "0     Fisher_observed            0.413950                  1.0          8   \n",
       "1     Fisher_observed            0.489554                  1.0         15   \n",
       "2     Fisher_observed            0.293439                  1.0          9   \n",
       "3     Fisher_observed            0.388602                  1.0          9   \n",
       "4     Fisher_observed            0.361455                  1.0          9   \n",
       "...               ...                 ...                  ...        ...   \n",
       "1203  Fisher_observed            0.247733                  1.0          2   \n",
       "1204  Fisher_observed            0.442825                  1.0          1   \n",
       "1205  Fisher_observed            0.414456                  1.0          1   \n",
       "1206  Fisher_observed            0.246445                  1.0          2   \n",
       "1207  Fisher_observed            0.244590                  1.0          2   \n",
       "\n",
       "      ...  Overlap  Final filter judgement  Expression_p-value  \\\n",
       "0     ...     True                    True            0.702076   \n",
       "1     ...     True                    True            0.694475   \n",
       "2     ...    False                   False            0.242469   \n",
       "3     ...     True                    True            0.231362   \n",
       "4     ...     True                   False            0.050680   \n",
       "...   ...      ...                     ...                 ...   \n",
       "1203  ...     True                   False                 NaN   \n",
       "1204  ...     True                   False                 NaN   \n",
       "1205  ...     True                   False                 NaN   \n",
       "1206  ...     True                   False                 NaN   \n",
       "1207  ...     True                   False                 NaN   \n",
       "\n",
       "      Expression_Fold Difference  Mutated patients w. expression data  \\\n",
       "0                       1.216133                                  4.0   \n",
       "1                       0.991591                                  8.0   \n",
       "2                       0.546071                                  6.0   \n",
       "3                       1.838830                                  2.0   \n",
       "4                       1.203716                                  5.0   \n",
       "...                          ...                                  ...   \n",
       "1203                         NaN                                  NaN   \n",
       "1204                         NaN                                  NaN   \n",
       "1205                         NaN                                  NaN   \n",
       "1206                         NaN                                  NaN   \n",
       "1207                         NaN                                  NaN   \n",
       "\n",
       "      Wild-type patients with expression data             file_name  chr  \\\n",
       "0                                       765.0  gc19_pc.promCore.bed   20   \n",
       "1                                       761.0  gc19_pc.promCore.bed   12   \n",
       "2                                       763.0  gc19_pc.promCore.bed    3   \n",
       "3                                       767.0  gc19_pc.promCore.bed    8   \n",
       "4                                       764.0  gc19_pc.promCore.bed    8   \n",
       "...                                       ...                   ...  ...   \n",
       "1203                                      NaN    smallrna.ncrna.bed   17   \n",
       "1204                                      NaN    smallrna.ncrna.bed    6   \n",
       "1205                                      NaN    smallrna.ncrna.bed    7   \n",
       "1206                                      NaN    smallrna.ncrna.bed   15   \n",
       "1207                                      NaN    smallrna.ncrna.bed    X   \n",
       "\n",
       "          start        end  \n",
       "0      32398909   32399274  \n",
       "1     120907452  120934115  \n",
       "2     193853733  193854133  \n",
       "3     103876327  103876578  \n",
       "4      29952713   29959655  \n",
       "...         ...        ...  \n",
       "1203   36908033   36908107  \n",
       "1204   27447452   27447524  \n",
       "1205  149295045  149295117  \n",
       "1206   25446469   25446551  \n",
       "1207  125864211  125864278  \n",
       "\n",
       "[1208 rows x 29 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_data = pd.read_csv('data/dataset_uncensored.csv')\n",
    "raw_data['start'] = raw_data['start'].astype(int)\n",
    "raw_data['end'] = raw_data['end'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlaps = check_overlaps(df, raw_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "promCore           28\n",
       "lncRNA_promCore    16\n",
       "5UTR               15\n",
       "enhancer            5\n",
       "lncRNA_ncrna        4\n",
       "3UTR                3\n",
       "Name: target, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlaps['target'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "arr_ids = []\n",
    "arr_ids = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/indel'))\n",
    "arr_ids  = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/snv_mnv'))\n",
    "# arr_ids = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/graylist'))\n",
    "arr_ids = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_tcga.controlled/final_consensus_12oct/tcga/indel'))\n",
    "arr_ids  = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_tcga.controlled/final_consensus_12oct/tcga/snv_mnv'))\n",
    "# arr_ids = arr_ids + (os.listdir('data/Rheinbay 2020/final_consensus_snv_indel_tcga.controlled/final_consensus_12oct/tcga/graylist'))\n",
    "arr_ids = [s for s in arr_ids if '.tbi' not in s]\n",
    "arr_ids_names = [s[:s.find(\".\")] for s in arr_ids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf_gzip(filename):\n",
    "    with gzip.open(filename, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith(b'##')]\n",
    "    df = pd.read_csv(\n",
    "        io.StringIO(b''.join(lines).decode()),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "                'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})\n",
    "    return df\n",
    "    \n",
    "read_vcf_gzip('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/indel/0009b464-b376-4fbc-8a56-da538269a02f.consensus.20161006.somatic.indel.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\enhancers.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\gc19_pc.3utr.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\gc19_pc.5utr.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\gc19_pc.cds.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\gc19_pc.promCore.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\gc19_pc.promDomain.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\gc19_pc.ss.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\lncrna.ncrna.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\lncrna.promCore.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\lncrna.promDomain.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\lncrna.ss.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\mirna.mat.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\mirna.pre.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\mirna.prom.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\smallrna.ncrna.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\supplementary_lncrna.ncrna.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\supplementary_lncrna.prom.bed',\n",
       " 'D:\\\\Masters Thesis\\\\GitHub Repo\\\\data\\\\gene-coding-and-non-coding-elements\\\\supplementary_lncrna.ss.bed']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_files = [val for sublist in [[os.path.join(i[0], j) for j in i[2]] for i in os.walk('D:\\Masters Thesis\\GitHub Repo\\data\\gene-coding-and-non-coding-elements')] for val in sublist]\n",
    "all_files = [s for s in all_files if 'allCombined' not in s]\n",
    "all_files = [s for s in all_files if 'mapFiles' not in s]\n",
    "all_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "all_df = pd.DataFrame()\n",
    "all_df['element'] = ''\n",
    "for file in all_files:\n",
    "    temp_df = pd.read_csv(file, sep='\\t', header=None)\n",
    "    result = re.search('elements\\\\\\\\(.*).bed', file)\n",
    "    temp_df['element'] = str(result.group(1))\n",
    "    all_df = pd.concat([all_df, temp_df])\n",
    "all_df.columns = ['element', 'chrom', 'chromStart','chromEnd', 'name', 'score', 'strand', 'thickStart', 'thickEnd', 'itemRgb', 'blockCount', 'blockSizes', 'blockStarts']\n",
    "all_df['chromStart'] = all_df['chromStart'].astype(int)\n",
    "all_df['chromEnd'] = all_df['chromEnd'].astype(int)\n",
    "all_df['score'] = all_df['score'].astype(int)\n",
    "all_df['chrom'] = all_df['chrom'].apply(lambda x: x.replace('chr', ''))\n",
    "\n",
    "all_df.rename({'chrom': 'chr', 'chromStart': 'start', 'chromEnd': 'end'}, inplace=True, axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found overlaps\n"
     ]
    }
   ],
   "source": [
    "overlaps = check_overlaps_and_annotate(raw_data, all_df, 'element')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>driver</th>\n",
       "      <th>data_source</th>\n",
       "      <th>found</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mut0</td>\n",
       "      <td>3</td>\n",
       "      <td>101578254</td>\n",
       "      <td>101578255</td>\n",
       "      <td>CAGTT</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>1</td>\n",
       "      <td>gc19_pc.3utr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mut1</td>\n",
       "      <td>3</td>\n",
       "      <td>101578285</td>\n",
       "      <td>101578286</td>\n",
       "      <td>GACCATTTGCCTT</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>1</td>\n",
       "      <td>gc19_pc.3utr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mut2</td>\n",
       "      <td>3</td>\n",
       "      <td>101578250</td>\n",
       "      <td>101578251</td>\n",
       "      <td>CTG</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>1</td>\n",
       "      <td>gc19_pc.3utr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mut3</td>\n",
       "      <td>17</td>\n",
       "      <td>48940015</td>\n",
       "      <td>48940016</td>\n",
       "      <td>CTAAAT</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>1</td>\n",
       "      <td>gc19_pc.3utr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mut4</td>\n",
       "      <td>17</td>\n",
       "      <td>48939986</td>\n",
       "      <td>48939987</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>1</td>\n",
       "      <td>gc19_pc.3utr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1081</th>\n",
       "      <td>mut1081</td>\n",
       "      <td>4</td>\n",
       "      <td>95496939</td>\n",
       "      <td>95496940</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>1</td>\n",
       "      <td>gc19_pc.ss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1082</th>\n",
       "      <td>mut1082</td>\n",
       "      <td>6</td>\n",
       "      <td>53026396</td>\n",
       "      <td>53026397</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>1</td>\n",
       "      <td>supplementary_lncrna.prom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1083</th>\n",
       "      <td>mut1083</td>\n",
       "      <td>2</td>\n",
       "      <td>220077160</td>\n",
       "      <td>220077161</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>1</td>\n",
       "      <td>gc19_pc.ss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1084</th>\n",
       "      <td>mut1084</td>\n",
       "      <td>16</td>\n",
       "      <td>31471209</td>\n",
       "      <td>31471210</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>1</td>\n",
       "      <td>gc19_pc.ss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1085</th>\n",
       "      <td>mut1085</td>\n",
       "      <td>18</td>\n",
       "      <td>59584293</td>\n",
       "      <td>59584294</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>1</td>\n",
       "      <td>supplementary_lncrna.ncrna</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>982 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id chr      start        end            ref alt  driver  \\\n",
       "0        mut0   3  101578254  101578255          CAGTT   C       1   \n",
       "1        mut1   3  101578285  101578286  GACCATTTGCCTT   G       1   \n",
       "2        mut2   3  101578250  101578251            CTG   C       1   \n",
       "3        mut3  17   48940015   48940016         CTAAAT   C       1   \n",
       "4        mut4  17   48939986   48939987             AT   A       1   \n",
       "...       ...  ..        ...        ...            ...  ..     ...   \n",
       "1081  mut1081   4   95496939   95496940              C   T       0   \n",
       "1082  mut1082   6   53026396   53026397              C   T       0   \n",
       "1083  mut1083   2  220077160  220077161              G   T       0   \n",
       "1084  mut1084  16   31471209   31471210              C   T       0   \n",
       "1085  mut1085  18   59584293   59584294              G   C       0   \n",
       "\n",
       "     data_source  found                  annotation  \n",
       "0           ICGC      1                gc19_pc.3utr  \n",
       "1           ICGC      1                gc19_pc.3utr  \n",
       "2           ICGC      1                gc19_pc.3utr  \n",
       "3           ICGC      1                gc19_pc.3utr  \n",
       "4           ICGC      1                gc19_pc.3utr  \n",
       "...          ...    ...                         ...  \n",
       "1081      COSMIC      1                  gc19_pc.ss  \n",
       "1082      COSMIC      1   supplementary_lncrna.prom  \n",
       "1083      COSMIC      1                  gc19_pc.ss  \n",
       "1084      COSMIC      1                  gc19_pc.ss  \n",
       "1085      COSMIC      1  supplementary_lncrna.ncrna  \n",
       "\n",
       "[982 rows x 10 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlaps[~overlaps['annotation'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found overlaps\n"
     ]
    }
   ],
   "source": [
    "overlaps1 = check_overlaps(all_df, raw_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "supplementary_lncrna.ncrna    184\n",
       "supplementary_lncrna.prom     170\n",
       "supplementary_lncrna.ss        90\n",
       "enhancers                       5\n",
       "gc19_pc.promCore                3\n",
       "gc19_pc.promDomain              3\n",
       "gc19_pc.ss                      3\n",
       "gc19_pc.3utr                    3\n",
       "gc19_pc.cds                     3\n",
       "gc19_pc.5utr                    3\n",
       "lncrna.ncrna                    2\n",
       "lncrna.promCore                 2\n",
       "lncrna.ss                       2\n",
       "mirna.mat                       2\n",
       "mirna.pre                       2\n",
       "mirna.prom                      2\n",
       "smallrna.ncrna                  2\n",
       "lncrna.promDomain               2\n",
       "Name: element, dtype: int64"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlaps1['element'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23159591"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def row_count(input):\n",
    "    with open(input) as f:\n",
    "        for i, l in enumerate(f):\n",
    "            pass\n",
    "    return i\n",
    "\n",
    "row_count('data/Rheinbay 2020/final_consensus_passonly.snv_mnv_indel.icgc.controlled.maf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1541"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arr_ids.index('6f25ca85-4b4a-420f-b0cc-5a4f65a97da3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/indel',\n",
       "      dtype='<U71')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.array('data/Rheinbay 2020/final_consensus_snv_indel_passonly_icgc.public/indel')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
