{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1052729</td>\n",
       "      <td>mut1831</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1227399</td>\n",
       "      <td>mut2068</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1234820</td>\n",
       "      <td>mut1462</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1256438</td>\n",
       "      <td>mut1952</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1342375</td>\n",
       "      <td>mut943</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3100</th>\n",
       "      <td>X</td>\n",
       "      <td>153006175</td>\n",
       "      <td>mut742</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3101</th>\n",
       "      <td>X</td>\n",
       "      <td>153174592</td>\n",
       "      <td>mut1820</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3102</th>\n",
       "      <td>X</td>\n",
       "      <td>153207408</td>\n",
       "      <td>mut1772</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3103</th>\n",
       "      <td>X</td>\n",
       "      <td>153698355</td>\n",
       "      <td>mut1225</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3104</th>\n",
       "      <td>Y</td>\n",
       "      <td>8101387</td>\n",
       "      <td>mut1702</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3105 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr      start       id ref alt\n",
       "0      1    1052729  mut1831   T   A\n",
       "1      1    1227399  mut2068   G   A\n",
       "2      1    1234820  mut1462   C   G\n",
       "3      1    1256438  mut1952   C   T\n",
       "4      1    1342375   mut943   A   G\n",
       "...   ..        ...      ...  ..  ..\n",
       "3100   X  153006175   mut742   T   G\n",
       "3101   X  153174592  mut1820   C   T\n",
       "3102   X  153207408  mut1772   T   G\n",
       "3103   X  153698355  mut1225   C   T\n",
       "3104   Y    8101387  mut1702   G   T\n",
       "\n",
       "[3105 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../data/extracted_data.csv')\n",
    "df[['chr', 'start', 'id', 'ref', 'alt']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "resdf = pd.DataFrame()\n",
    "idx = 0\n",
    "while idx < len(df):\n",
    "    data_str = '{ \"variants\" : ['\n",
    "    while True and idx < len(df):\n",
    "        data_str = data_str + '\"{}  {}  {}  {}  {}  .   .   .\",'.format(df.iloc[idx]['chr'], df.iloc[idx]['start'], df.iloc[idx]['id'], df.iloc[idx]['ref'], df.iloc[idx]['alt'])\n",
    "        idx = idx + 1\n",
    "        if idx % 200 == 0:\n",
    "            break\n",
    "    data_str = data_str[:-1]\n",
    "    data_str = data_str + ']}'\n",
    "    server = \"https://grch37.rest.ensembl.org\"\n",
    "    ext = \"/vep/homo_sapiens/region?GO=True&SpliceAI=True&dbscSNV=True&protein=True&uniprot=True\"\n",
    "    headers={ \"Content-Type\" : \"application/json\", \"Accept\" : \"application/json\"}\n",
    "    r = requests.post(server+ext, headers=headers, data=data_str) # '{ \"variants\" : [\"21  26960070  rs116645811 G A . . .\", \"21  26965148  rs1135638 G A . . .\" ] }'\n",
    "    \n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        sys.exit()\n",
    "    \n",
    "    decoded = r.json()\n",
    "    print(repr(decoded))\n",
    "    resdf = pd.concat([resdf, pd.DataFrame(decoded)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_string</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>regulatory_feature_consequences</th>\n",
       "      <th>strand</th>\n",
       "      <th>seq_region_name</th>\n",
       "      <th>input</th>\n",
       "      <th>transcript_consequences</th>\n",
       "      <th>id</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>end</th>\n",
       "      <th>start</th>\n",
       "      <th>colocated_variants</th>\n",
       "      <th>motif_feature_consequences</th>\n",
       "      <th>intergenic_consequences</th>\n",
       "      <th>minimised</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T/A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[{'consequence_terms': ['regulatory_region_var...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1 1052729 mut1831 T A . . .</td>\n",
       "      <td>[{'transcript_id': 'ENST00000294576', 'uniparc...</td>\n",
       "      <td>mut1831</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1052729</td>\n",
       "      <td>1052729</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G/A</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>[{'variant_allele': 'A', 'regulatory_feature_i...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1 1227399 mut2068 G A . . .</td>\n",
       "      <td>[{'hgnc_id': 10601, 'impact': 'MODIFIER', 'tra...</td>\n",
       "      <td>mut2068</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1227399</td>\n",
       "      <td>1227399</td>\n",
       "      <td>[{'phenotype_or_disease': 1, 'strand': 1, 'all...</td>\n",
       "      <td>[{'motif_pos': 1, 'impact': 'MODIFIER', 'high_...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C/G</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>[{'impact': 'MODIFIER', 'variant_allele': 'G',...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1 1234820 mut1462 C G . . .</td>\n",
       "      <td>[{'strand': -1, 'gene_symbol_source': 'HGNC', ...</td>\n",
       "      <td>mut1462</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1234820</td>\n",
       "      <td>1234820</td>\n",
       "      <td>[{'end': 1234820, 'id': 'COSV61222486', 'seq_r...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C/T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1 1256438 mut1952 C T . . .</td>\n",
       "      <td>[{'transcript_id': 'ENST00000323275', 'hgnc_id...</td>\n",
       "      <td>mut1952</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1256438</td>\n",
       "      <td>1256438</td>\n",
       "      <td>[{'end': 1256438, 'id': 'COSV59672305', 'seq_r...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A/G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>[{'impact': 'MODIFIER', 'regulatory_feature_id...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1 1342375 mut943 A G . . .</td>\n",
       "      <td>[{'amino_acids': 'C/R', 'codons': 'Tgc/Cgc', '...</td>\n",
       "      <td>mut943</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1342375</td>\n",
       "      <td>1342375</td>\n",
       "      <td>[{'var_synonyms': {'COSMIC': ['COSM4703202']},...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>T/G</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "      <td>[{'regulatory_feature_id': 'ENSR00001770375', ...</td>\n",
       "      <td>1</td>\n",
       "      <td>X</td>\n",
       "      <td>X 153006175 mut742 T G . . .</td>\n",
       "      <td>[{'biotype': 'protein_coding', 'gene_symbol_so...</td>\n",
       "      <td>mut742</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>153006175</td>\n",
       "      <td>153006175</td>\n",
       "      <td>[{'seq_region_name': 'X', 'id': 'COSV99497660'...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>C/T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>[{'regulatory_feature_id': 'ENSR00002105717', ...</td>\n",
       "      <td>1</td>\n",
       "      <td>X</td>\n",
       "      <td>X 153174592 mut1820 C T . . .</td>\n",
       "      <td>[{'gene_symbol_source': 'HGNC', 'variant_allel...</td>\n",
       "      <td>mut1820</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>153174592</td>\n",
       "      <td>153174592</td>\n",
       "      <td>[{'allele_string': 'COSMIC_MUTATION', 'end': 1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>T/G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>[{'regulatory_feature_id': 'ENSR00000424671', ...</td>\n",
       "      <td>1</td>\n",
       "      <td>X</td>\n",
       "      <td>X 153207408 mut1772 T G . . .</td>\n",
       "      <td>[{'protein_start': 241, 'hgnc_id': 9959, 'cons...</td>\n",
       "      <td>mut1772</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>153207408</td>\n",
       "      <td>153207408</td>\n",
       "      <td>[{'var_synonyms': {'COSMIC': ['COSM6364490']},...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>C/T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>X</td>\n",
       "      <td>X 153698355 mut1225 C T . . .</td>\n",
       "      <td>[{'gene_symbol_source': 'HGNC', 'impact': 'MOD...</td>\n",
       "      <td>mut1225</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>153698355</td>\n",
       "      <td>153698355</td>\n",
       "      <td>[{'phenotype_or_disease': 1, 'id': 'COSV105281...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>G/T</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y 8101387 mut1702 G T . . .</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mut1702</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>8101387</td>\n",
       "      <td>8101387</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'variant_allele': 'T', 'impact': 'MODIFIER',...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3056 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    allele_string             most_severe_consequence  \\\n",
       "0             T/A               upstream_gene_variant   \n",
       "1             G/A                 3_prime_UTR_variant   \n",
       "2             C/G  non_coding_transcript_exon_variant   \n",
       "3             C/T                         stop_gained   \n",
       "4             A/G                    missense_variant   \n",
       "..            ...                                 ...   \n",
       "100           T/G                splice_donor_variant   \n",
       "101           C/T                    missense_variant   \n",
       "102           T/G                    missense_variant   \n",
       "103           C/T                    missense_variant   \n",
       "104           G/T                  intergenic_variant   \n",
       "\n",
       "                       regulatory_feature_consequences  strand  \\\n",
       "0    [{'consequence_terms': ['regulatory_region_var...       1   \n",
       "1    [{'variant_allele': 'A', 'regulatory_feature_i...       1   \n",
       "2    [{'impact': 'MODIFIER', 'variant_allele': 'G',...       1   \n",
       "3                                                  NaN       1   \n",
       "4    [{'impact': 'MODIFIER', 'regulatory_feature_id...       1   \n",
       "..                                                 ...     ...   \n",
       "100  [{'regulatory_feature_id': 'ENSR00001770375', ...       1   \n",
       "101  [{'regulatory_feature_id': 'ENSR00002105717', ...       1   \n",
       "102  [{'regulatory_feature_id': 'ENSR00000424671', ...       1   \n",
       "103                                                NaN       1   \n",
       "104                                                NaN       1   \n",
       "\n",
       "    seq_region_name                          input  \\\n",
       "0                 1    1 1052729 mut1831 T A . . .   \n",
       "1                 1    1 1227399 mut2068 G A . . .   \n",
       "2                 1    1 1234820 mut1462 C G . . .   \n",
       "3                 1    1 1256438 mut1952 C T . . .   \n",
       "4                 1     1 1342375 mut943 A G . . .   \n",
       "..              ...                            ...   \n",
       "100               X   X 153006175 mut742 T G . . .   \n",
       "101               X  X 153174592 mut1820 C T . . .   \n",
       "102               X  X 153207408 mut1772 T G . . .   \n",
       "103               X  X 153698355 mut1225 C T . . .   \n",
       "104               Y    Y 8101387 mut1702 G T . . .   \n",
       "\n",
       "                               transcript_consequences       id assembly_name  \\\n",
       "0    [{'transcript_id': 'ENST00000294576', 'uniparc...  mut1831        GRCh37   \n",
       "1    [{'hgnc_id': 10601, 'impact': 'MODIFIER', 'tra...  mut2068        GRCh37   \n",
       "2    [{'strand': -1, 'gene_symbol_source': 'HGNC', ...  mut1462        GRCh37   \n",
       "3    [{'transcript_id': 'ENST00000323275', 'hgnc_id...  mut1952        GRCh37   \n",
       "4    [{'amino_acids': 'C/R', 'codons': 'Tgc/Cgc', '...   mut943        GRCh37   \n",
       "..                                                 ...      ...           ...   \n",
       "100  [{'biotype': 'protein_coding', 'gene_symbol_so...   mut742        GRCh37   \n",
       "101  [{'gene_symbol_source': 'HGNC', 'variant_allel...  mut1820        GRCh37   \n",
       "102  [{'protein_start': 241, 'hgnc_id': 9959, 'cons...  mut1772        GRCh37   \n",
       "103  [{'gene_symbol_source': 'HGNC', 'impact': 'MOD...  mut1225        GRCh37   \n",
       "104                                                NaN  mut1702        GRCh37   \n",
       "\n",
       "           end      start                                 colocated_variants  \\\n",
       "0      1052729    1052729                                                NaN   \n",
       "1      1227399    1227399  [{'phenotype_or_disease': 1, 'strand': 1, 'all...   \n",
       "2      1234820    1234820  [{'end': 1234820, 'id': 'COSV61222486', 'seq_r...   \n",
       "3      1256438    1256438  [{'end': 1256438, 'id': 'COSV59672305', 'seq_r...   \n",
       "4      1342375    1342375  [{'var_synonyms': {'COSMIC': ['COSM4703202']},...   \n",
       "..         ...        ...                                                ...   \n",
       "100  153006175  153006175  [{'seq_region_name': 'X', 'id': 'COSV99497660'...   \n",
       "101  153174592  153174592  [{'allele_string': 'COSMIC_MUTATION', 'end': 1...   \n",
       "102  153207408  153207408  [{'var_synonyms': {'COSMIC': ['COSM6364490']},...   \n",
       "103  153698355  153698355  [{'phenotype_or_disease': 1, 'id': 'COSV105281...   \n",
       "104    8101387    8101387                                                NaN   \n",
       "\n",
       "                            motif_feature_consequences  \\\n",
       "0                                                  NaN   \n",
       "1    [{'motif_pos': 1, 'impact': 'MODIFIER', 'high_...   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "100                                                NaN   \n",
       "101                                                NaN   \n",
       "102                                                NaN   \n",
       "103                                                NaN   \n",
       "104                                                NaN   \n",
       "\n",
       "                               intergenic_consequences  minimised  \n",
       "0                                                  NaN        NaN  \n",
       "1                                                  NaN        NaN  \n",
       "2                                                  NaN        NaN  \n",
       "3                                                  NaN        NaN  \n",
       "4                                                  NaN        NaN  \n",
       "..                                                 ...        ...  \n",
       "100                                                NaN        NaN  \n",
       "101                                                NaN        NaN  \n",
       "102                                                NaN        NaN  \n",
       "103                                                NaN        NaN  \n",
       "104  [{'variant_allele': 'T', 'impact': 'MODIFIER',...        NaN  \n",
       "\n",
       "[3056 rows x 15 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resdf#.to_pickle('vep_api_output.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1.5 min for 1 call (300 mutations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input</th>\n",
       "      <th>seq_region_name</th>\n",
       "      <th>end</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>most_severe_consequence</th>\n",
       "      <th>start</th>\n",
       "      <th>regulatory_feature_consequences</th>\n",
       "      <th>transcript_consequences</th>\n",
       "      <th>allele_string</th>\n",
       "      <th>id</th>\n",
       "      <th>colocated_variants</th>\n",
       "      <th>motif_feature_consequences</th>\n",
       "      <th>intergenic_consequences</th>\n",
       "      <th>minimised</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1 1052729 mut1831 T A . . .</td>\n",
       "      <td>1</td>\n",
       "      <td>1052729</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>1052729</td>\n",
       "      <td>[{'variant_allele': 'A', 'biotype': 'promoter'...</td>\n",
       "      <td>[{'gene_symbol_source': 'HGNC', 'strand': -1, ...</td>\n",
       "      <td>T/A</td>\n",
       "      <td>mut1831</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1 1227399 mut2068 G A . . .</td>\n",
       "      <td>1</td>\n",
       "      <td>1227399</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>1227399</td>\n",
       "      <td>[{'regulatory_feature_id': 'ENSR00000249817', ...</td>\n",
       "      <td>[{'protein_id': 'ENSP00000321594', 'uniparc': ...</td>\n",
       "      <td>G/A</td>\n",
       "      <td>mut2068</td>\n",
       "      <td>[{'allele_string': 'COSMIC_MUTATION', 'id': 'C...</td>\n",
       "      <td>[{'strand': 1, 'motif_name': 'ENSPFM0042', 'va...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1 1234820 mut1462 C G . . .</td>\n",
       "      <td>1</td>\n",
       "      <td>1234820</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>1234820</td>\n",
       "      <td>[{'regulatory_feature_id': 'ENSR00000918547', ...</td>\n",
       "      <td>[{'protein_id': 'ENSP00000321139', 'go': [{'go...</td>\n",
       "      <td>C/G</td>\n",
       "      <td>mut1462</td>\n",
       "      <td>[{'start': 1234820, 'end': 1234820, 'seq_regio...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1 1256438 mut1952 C T . . .</td>\n",
       "      <td>1</td>\n",
       "      <td>1256438</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>1256438</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'cdna_start': 77, 'transcript_id': 'ENST0000...</td>\n",
       "      <td>C/T</td>\n",
       "      <td>mut1952</td>\n",
       "      <td>[{'id': 'COSV59672305', 'allele_string': 'COSM...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1 1342375 mut943 A G . . .</td>\n",
       "      <td>1</td>\n",
       "      <td>1342375</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1342375</td>\n",
       "      <td>[{'variant_allele': 'G', 'biotype': 'promoter'...</td>\n",
       "      <td>[{'gene_symbol_source': 'HGNC', 'sift_predicti...</td>\n",
       "      <td>A/G</td>\n",
       "      <td>mut943</td>\n",
       "      <td>[{'allele_string': 'COSMIC_MUTATION', 'id': 'C...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>X 153006175 mut742 T G . . .</td>\n",
       "      <td>X</td>\n",
       "      <td>153006175</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>splice_donor_variant</td>\n",
       "      <td>153006175</td>\n",
       "      <td>[{'consequence_terms': ['regulatory_region_var...</td>\n",
       "      <td>[{'variant_allele': 'G', 'uniparc': ['UPI00000...</td>\n",
       "      <td>T/G</td>\n",
       "      <td>mut742</td>\n",
       "      <td>[{'var_synonyms': {'COSMIC': ['COSM8997094']},...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>X 153174592 mut1820 C T . . .</td>\n",
       "      <td>X</td>\n",
       "      <td>153174592</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>153174592</td>\n",
       "      <td>[{'biotype': 'open_chromatin_region', 'consequ...</td>\n",
       "      <td>[{'distance': 1975, 'protein_id': 'ENSP0000033...</td>\n",
       "      <td>C/T</td>\n",
       "      <td>mut1820</td>\n",
       "      <td>[{'start': 153174592, 'strand': 1, 'allele_str...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>X 153207408 mut1772 T G . . .</td>\n",
       "      <td>X</td>\n",
       "      <td>153207408</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>153207408</td>\n",
       "      <td>[{'impact': 'MODIFIER', 'regulatory_feature_id...</td>\n",
       "      <td>[{'polyphen_prediction': 'possibly_damaging', ...</td>\n",
       "      <td>T/G</td>\n",
       "      <td>mut1772</td>\n",
       "      <td>[{'phenotype_or_disease': 1, 'id': 'COSV641696...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>X 153698355 mut1225 C T . . .</td>\n",
       "      <td>X</td>\n",
       "      <td>153698355</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>153698355</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'hgnc_id': 9101, 'gene_symbol': 'PLXNA3', 'g...</td>\n",
       "      <td>C/T</td>\n",
       "      <td>mut1225</td>\n",
       "      <td>[{'var_synonyms': {'COSMIC': ['COSM9741561']},...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>Y 8101387 mut1702 G T . . .</td>\n",
       "      <td>Y</td>\n",
       "      <td>8101387</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>8101387</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G/T</td>\n",
       "      <td>mut1702</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'impact': 'MODIFIER', 'consequence_terms': [...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3056 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             input seq_region_name        end assembly_name  \\\n",
       "0      1 1052729 mut1831 T A . . .               1    1052729        GRCh37   \n",
       "1      1 1227399 mut2068 G A . . .               1    1227399        GRCh37   \n",
       "2      1 1234820 mut1462 C G . . .               1    1234820        GRCh37   \n",
       "3      1 1256438 mut1952 C T . . .               1    1256438        GRCh37   \n",
       "4       1 1342375 mut943 A G . . .               1    1342375        GRCh37   \n",
       "..                             ...             ...        ...           ...   \n",
       "100   X 153006175 mut742 T G . . .               X  153006175        GRCh37   \n",
       "101  X 153174592 mut1820 C T . . .               X  153174592        GRCh37   \n",
       "102  X 153207408 mut1772 T G . . .               X  153207408        GRCh37   \n",
       "103  X 153698355 mut1225 C T . . .               X  153698355        GRCh37   \n",
       "104    Y 8101387 mut1702 G T . . .               Y    8101387        GRCh37   \n",
       "\n",
       "     strand             most_severe_consequence      start  \\\n",
       "0         1               upstream_gene_variant    1052729   \n",
       "1         1                 3_prime_UTR_variant    1227399   \n",
       "2         1  non_coding_transcript_exon_variant    1234820   \n",
       "3         1                         stop_gained    1256438   \n",
       "4         1                    missense_variant    1342375   \n",
       "..      ...                                 ...        ...   \n",
       "100       1                splice_donor_variant  153006175   \n",
       "101       1                    missense_variant  153174592   \n",
       "102       1                    missense_variant  153207408   \n",
       "103       1                    missense_variant  153698355   \n",
       "104       1                  intergenic_variant    8101387   \n",
       "\n",
       "                       regulatory_feature_consequences  \\\n",
       "0    [{'variant_allele': 'A', 'biotype': 'promoter'...   \n",
       "1    [{'regulatory_feature_id': 'ENSR00000249817', ...   \n",
       "2    [{'regulatory_feature_id': 'ENSR00000918547', ...   \n",
       "3                                                  NaN   \n",
       "4    [{'variant_allele': 'G', 'biotype': 'promoter'...   \n",
       "..                                                 ...   \n",
       "100  [{'consequence_terms': ['regulatory_region_var...   \n",
       "101  [{'biotype': 'open_chromatin_region', 'consequ...   \n",
       "102  [{'impact': 'MODIFIER', 'regulatory_feature_id...   \n",
       "103                                                NaN   \n",
       "104                                                NaN   \n",
       "\n",
       "                               transcript_consequences allele_string       id  \\\n",
       "0    [{'gene_symbol_source': 'HGNC', 'strand': -1, ...           T/A  mut1831   \n",
       "1    [{'protein_id': 'ENSP00000321594', 'uniparc': ...           G/A  mut2068   \n",
       "2    [{'protein_id': 'ENSP00000321139', 'go': [{'go...           C/G  mut1462   \n",
       "3    [{'cdna_start': 77, 'transcript_id': 'ENST0000...           C/T  mut1952   \n",
       "4    [{'gene_symbol_source': 'HGNC', 'sift_predicti...           A/G   mut943   \n",
       "..                                                 ...           ...      ...   \n",
       "100  [{'variant_allele': 'G', 'uniparc': ['UPI00000...           T/G   mut742   \n",
       "101  [{'distance': 1975, 'protein_id': 'ENSP0000033...           C/T  mut1820   \n",
       "102  [{'polyphen_prediction': 'possibly_damaging', ...           T/G  mut1772   \n",
       "103  [{'hgnc_id': 9101, 'gene_symbol': 'PLXNA3', 'g...           C/T  mut1225   \n",
       "104                                                NaN           G/T  mut1702   \n",
       "\n",
       "                                    colocated_variants  \\\n",
       "0                                                  NaN   \n",
       "1    [{'allele_string': 'COSMIC_MUTATION', 'id': 'C...   \n",
       "2    [{'start': 1234820, 'end': 1234820, 'seq_regio...   \n",
       "3    [{'id': 'COSV59672305', 'allele_string': 'COSM...   \n",
       "4    [{'allele_string': 'COSMIC_MUTATION', 'id': 'C...   \n",
       "..                                                 ...   \n",
       "100  [{'var_synonyms': {'COSMIC': ['COSM8997094']},...   \n",
       "101  [{'start': 153174592, 'strand': 1, 'allele_str...   \n",
       "102  [{'phenotype_or_disease': 1, 'id': 'COSV641696...   \n",
       "103  [{'var_synonyms': {'COSMIC': ['COSM9741561']},...   \n",
       "104                                                NaN   \n",
       "\n",
       "                            motif_feature_consequences  \\\n",
       "0                                                  NaN   \n",
       "1    [{'strand': 1, 'motif_name': 'ENSPFM0042', 'va...   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "100                                                NaN   \n",
       "101                                                NaN   \n",
       "102                                                NaN   \n",
       "103                                                NaN   \n",
       "104                                                NaN   \n",
       "\n",
       "                               intergenic_consequences  minimised  \n",
       "0                                                  NaN        NaN  \n",
       "1                                                  NaN        NaN  \n",
       "2                                                  NaN        NaN  \n",
       "3                                                  NaN        NaN  \n",
       "4                                                  NaN        NaN  \n",
       "..                                                 ...        ...  \n",
       "100                                                NaN        NaN  \n",
       "101                                                NaN        NaN  \n",
       "102                                                NaN        NaN  \n",
       "103                                                NaN        NaN  \n",
       "104  [{'impact': 'MODIFIER', 'consequence_terms': [...        NaN  \n",
       "\n",
       "[3056 rows x 15 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# resdf = pd.read_csv('../data/VEP/vep_api_output.csv')\n",
    "resdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'consequence_terms': ['3_prime_UTR_variant'],\n",
       "  'biotype': 'protein_coding',\n",
       "  'gene_id': 'ENSG00000162572',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'strand': 1,\n",
       "  'cdna_end': 2674,\n",
       "  'gene_symbol': 'SCNN1D',\n",
       "  'go': [{'description': 'sodium_channel_activity', 'go_term': 'GO:0005272'},\n",
       "   {'go_term': 'GO:0005515', 'description': 'protein_binding'},\n",
       "   {'go_term': 'GO:0005886', 'description': 'plasma_membrane'},\n",
       "   {'description': 'sodium_ion_transport', 'go_term': 'GO:0006814'},\n",
       "   {'go_term': 'GO:0015280',\n",
       "    'description': 'ligand-gated_sodium_channel_activity'},\n",
       "   {'description': 'membrane', 'go_term': 'GO:0016020'},\n",
       "   {'go_term': 'GO:0016021', 'description': 'integral_component_of_membrane'},\n",
       "   {'go_term': 'GO:0034220', 'description': 'ion_transmembrane_transport'},\n",
       "   {'description': 'response_to_stimulus', 'go_term': 'GO:0050896'},\n",
       "   {'description': 'sensory_perception_of_taste', 'go_term': 'GO:0050909'},\n",
       "   {'go_term': 'GO:0055085', 'description': 'transmembrane_transport'}],\n",
       "  'variant_allele': 'A',\n",
       "  'cdna_start': 2674,\n",
       "  'impact': 'MODIFIER',\n",
       "  'hgnc_id': 10601,\n",
       "  'transcript_id': 'ENST00000325425'},\n",
       " {'gene_symbol': 'SCNN1D',\n",
       "  'go': [{'go_term': 'GO:0005272', 'description': 'sodium_channel_activity'},\n",
       "   {'go_term': 'GO:0005515', 'description': 'protein_binding'},\n",
       "   {'go_term': 'GO:0005886', 'description': 'plasma_membrane'},\n",
       "   {'go_term': 'GO:0006814', 'description': 'sodium_ion_transport'},\n",
       "   {'go_term': 'GO:0015280',\n",
       "    'description': 'ligand-gated_sodium_channel_activity'},\n",
       "   {'go_term': 'GO:0016020', 'description': 'membrane'},\n",
       "   {'description': 'integral_component_of_membrane', 'go_term': 'GO:0016021'},\n",
       "   {'go_term': 'GO:0034220', 'description': 'ion_transmembrane_transport'},\n",
       "   {'go_term': 'GO:0050896', 'description': 'response_to_stimulus'},\n",
       "   {'description': 'sensory_perception_of_taste', 'go_term': 'GO:0050909'},\n",
       "   {'description': 'transmembrane_transport', 'go_term': 'GO:0055085'}],\n",
       "  'variant_allele': 'A',\n",
       "  'cdna_start': 3470,\n",
       "  'hgnc_id': 10601,\n",
       "  'impact': 'MODIFIER',\n",
       "  'transcript_id': 'ENST00000338555',\n",
       "  'consequence_terms': ['3_prime_UTR_variant'],\n",
       "  'biotype': 'protein_coding',\n",
       "  'gene_id': 'ENSG00000162572',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'cdna_end': 3470,\n",
       "  'strand': 1},\n",
       " {'consequence_terms': ['downstream_gene_variant'],\n",
       "  'biotype': 'protein_coding',\n",
       "  'gene_id': 'ENSG00000131584',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'distance': 365,\n",
       "  'strand': -1,\n",
       "  'gene_symbol': 'ACAP3',\n",
       "  'go': [{'go_term': 'GO:0005515', 'description': 'protein_binding'},\n",
       "   {'description': 'ARF_GTPase_activator_activity', 'go_term': 'GO:0008060'},\n",
       "   {'go_term': 'GO:0008270', 'description': 'zinc_ion_binding'},\n",
       "   {'description': 'regulation_of_ARF_GTPase_activity',\n",
       "    'go_term': 'GO:0032312'}],\n",
       "  'variant_allele': 'A',\n",
       "  'hgnc_id': 16754,\n",
       "  'impact': 'MODIFIER',\n",
       "  'transcript_id': 'ENST00000353662'},\n",
       " {'gene_symbol': 'ACAP3',\n",
       "  'go': [{'go_term': 'GO:0005515', 'description': 'protein_binding'},\n",
       "   {'go_term': 'GO:0008060', 'description': 'ARF_GTPase_activator_activity'},\n",
       "   {'description': 'zinc_ion_binding', 'go_term': 'GO:0008270'},\n",
       "   {'go_term': 'GO:0032312',\n",
       "    'description': 'regulation_of_ARF_GTPase_activity'}],\n",
       "  'variant_allele': 'A',\n",
       "  'hgnc_id': 16754,\n",
       "  'impact': 'MODIFIER',\n",
       "  'transcript_id': 'ENST00000354700',\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'biotype': 'protein_coding',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'gene_id': 'ENSG00000131584',\n",
       "  'distance': 357,\n",
       "  'strand': -1},\n",
       " {'transcript_id': 'ENST00000379037',\n",
       "  'impact': 'MODIFIER',\n",
       "  'hgnc_id': 16754,\n",
       "  'variant_allele': 'A',\n",
       "  'gene_symbol': 'ACAP3',\n",
       "  'strand': -1,\n",
       "  'distance': 4417,\n",
       "  'gene_id': 'ENSG00000131584',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'biotype': 'processed_transcript'},\n",
       " {'gene_symbol': 'SCNN1D',\n",
       "  'go': [{'go_term': 'GO:0005272', 'description': 'sodium_channel_activity'},\n",
       "   {'description': 'sodium_ion_transport', 'go_term': 'GO:0006814'},\n",
       "   {'description': 'membrane', 'go_term': 'GO:0016020'}],\n",
       "  'variant_allele': 'A',\n",
       "  'hgnc_id': 10601,\n",
       "  'impact': 'MODIFIER',\n",
       "  'flags': ['cds_start_NF', 'cds_end_NF'],\n",
       "  'transcript_id': 'ENST00000379099',\n",
       "  'biotype': 'protein_coding',\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'gene_id': 'ENSG00000162572',\n",
       "  'distance': 1076,\n",
       "  'strand': 1},\n",
       " {'distance': 646,\n",
       "  'strand': 1,\n",
       "  'biotype': 'nonsense_mediated_decay',\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'gene_id': 'ENSG00000162572',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'hgnc_id': 10601,\n",
       "  'impact': 'MODIFIER',\n",
       "  'transcript_id': 'ENST00000379101',\n",
       "  'gene_symbol': 'SCNN1D',\n",
       "  'variant_allele': 'A'},\n",
       " {'strand': 1,\n",
       "  'cdna_end': 3044,\n",
       "  'biotype': 'protein_coding',\n",
       "  'consequence_terms': ['3_prime_UTR_variant'],\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'gene_id': 'ENSG00000162572',\n",
       "  'hgnc_id': 10601,\n",
       "  'impact': 'MODIFIER',\n",
       "  'transcript_id': 'ENST00000379116',\n",
       "  'go': [{'go_term': 'GO:0005272', 'description': 'sodium_channel_activity'},\n",
       "   {'description': 'sodium_ion_transport', 'go_term': 'GO:0006814'},\n",
       "   {'description': 'ligand-gated_sodium_channel_activity',\n",
       "    'go_term': 'GO:0015280'},\n",
       "   {'description': 'membrane', 'go_term': 'GO:0016020'},\n",
       "   {'description': 'integral_component_of_membrane', 'go_term': 'GO:0016021'}],\n",
       "  'gene_symbol': 'SCNN1D',\n",
       "  'cdna_start': 3044,\n",
       "  'variant_allele': 'A'},\n",
       " {'gene_id': 'ENSG00000162572',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'consequence_terms': ['3_prime_UTR_variant'],\n",
       "  'biotype': 'protein_coding',\n",
       "  'cdna_end': 2493,\n",
       "  'strand': 1,\n",
       "  'variant_allele': 'A',\n",
       "  'cdna_start': 2493,\n",
       "  'go': [{'description': 'sodium_channel_activity', 'go_term': 'GO:0005272'},\n",
       "   {'go_term': 'GO:0005515', 'description': 'protein_binding'},\n",
       "   {'description': 'plasma_membrane', 'go_term': 'GO:0005886'},\n",
       "   {'go_term': 'GO:0006814', 'description': 'sodium_ion_transport'},\n",
       "   {'go_term': 'GO:0015280',\n",
       "    'description': 'ligand-gated_sodium_channel_activity'},\n",
       "   {'description': 'membrane', 'go_term': 'GO:0016020'},\n",
       "   {'go_term': 'GO:0016021', 'description': 'integral_component_of_membrane'},\n",
       "   {'go_term': 'GO:0034220', 'description': 'ion_transmembrane_transport'},\n",
       "   {'description': 'response_to_stimulus', 'go_term': 'GO:0050896'},\n",
       "   {'go_term': 'GO:0050909', 'description': 'sensory_perception_of_taste'},\n",
       "   {'go_term': 'GO:0055085', 'description': 'transmembrane_transport'}],\n",
       "  'gene_symbol': 'SCNN1D',\n",
       "  'transcript_id': 'ENST00000400928',\n",
       "  'hgnc_id': 10601,\n",
       "  'impact': 'MODIFIER'},\n",
       " {'gene_symbol': 'ACAP3',\n",
       "  'variant_allele': 'A',\n",
       "  'hgnc_id': 16754,\n",
       "  'impact': 'MODIFIER',\n",
       "  'transcript_id': 'ENST00000467278',\n",
       "  'biotype': 'retained_intron',\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'gene_id': 'ENSG00000131584',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'distance': 851,\n",
       "  'strand': -1},\n",
       " {'transcript_id': 'ENST00000470659',\n",
       "  'impact': 'MODIFIER',\n",
       "  'hgnc_id': 16754,\n",
       "  'variant_allele': 'A',\n",
       "  'gene_symbol': 'ACAP3',\n",
       "  'strand': -1,\n",
       "  'distance': 2480,\n",
       "  'gene_id': 'ENSG00000131584',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'biotype': 'retained_intron'},\n",
       " {'gene_id': 'ENSG00000131584',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'biotype': 'retained_intron',\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'strand': -1,\n",
       "  'distance': 4447,\n",
       "  'variant_allele': 'A',\n",
       "  'gene_symbol': 'ACAP3',\n",
       "  'transcript_id': 'ENST00000476572',\n",
       "  'hgnc_id': 16754,\n",
       "  'impact': 'MODIFIER'},\n",
       " {'impact': 'MODIFIER',\n",
       "  'hgnc_id': 16754,\n",
       "  'transcript_id': 'ENST00000492936',\n",
       "  'gene_symbol': 'ACAP3',\n",
       "  'variant_allele': 'A',\n",
       "  'distance': 371,\n",
       "  'strand': -1,\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'biotype': 'retained_intron',\n",
       "  'gene_id': 'ENSG00000131584',\n",
       "  'gene_symbol_source': 'HGNC'}]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resdf.iloc[1]['transcript_consequences']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_allele</th>\n",
       "      <th>biotype</th>\n",
       "      <th>impact</th>\n",
       "      <th>consequence_terms</th>\n",
       "      <th>regulatory_feature_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>promoter</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>[regulatory_region_variant]</td>\n",
       "      <td>ENSR00000000136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  variant_allele   biotype    impact            consequence_terms  \\\n",
       "0              A  promoter  MODIFIER  [regulatory_region_variant]   \n",
       "\n",
       "  regulatory_feature_id  \n",
       "0       ENSR00000000136  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# resdf#.columns\n",
    "import json\n",
    "pd.json_normalize(resdf.regulatory_feature_consequences.iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             consequence_terms regulatory_feature_id   biotype    impact  \\\n",
      "0  [regulatory_region_variant]       ENSR00000000136  promoter  MODIFIER   \n",
      "\n",
      "  variant_allele  \n",
      "0              A  \n"
     ]
    }
   ],
   "source": [
    "# resdf[[VEP_COLUMNS]] = np.nan\n",
    "# resdf\n",
    "for index, row in resdf.iterrows():\n",
    "    rfc = pd.json_normalize(row['regulatory_feature_consequences'])\n",
    "    resdf.at[index, rfc.iloc[0]['biotype']] = 1\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "VEP_COLUMNS = ['ada_score', 'rf_score',\n",
    "'ENSP', 'UNIPARC', 'GO',\n",
    "'LOEUF', 'SpliceAI_pred_DP_AG', 'SpliceAI_pred_DP_AL',\n",
    "'SpliceAI_pred_DP_DG', 'SpliceAI_pred_DP_DL', 'SpliceAI_pred_DS_AG',\n",
    "'SpliceAI_pred_DS_AL', 'SpliceAI_pred_DS_DG', 'SpliceAI_pred_DS_DL',\n",
    "'3_prime_UTR_variant', '5_prime_UTR_variant',\n",
    "'NMD_transcript_variant', \n",
    "'downstream_gene_variant',\n",
    "'intergenic_variant', 'intron_variant',\n",
    "'non_coding_transcript_exon_variant', 'non_coding_transcript_variant',\n",
    "'regulatory_region_variant',\n",
    "'splice_donor_variant', 'splice_polypyrimidine_tract_variant',\n",
    "'upstream_gene_variant', 'MODIFIER', 'MotifFeature',\n",
    "'RegulatoryFeature', 'Transcript', 'CTCF_binding_site',\n",
    "'enhancer', 'nonsense_mediated_decay',\n",
    "'processed_pseudogene', 'processed_transcript', 'promoter',\n",
    "'protein_coding', 'retained_intron',\n",
    "'unprocessed_pseudogene',\n",
    "'splice_acceptor_variant', 'splice_donor_region_variant', 'splice_donor_5th_base_variant',\n",
    "'missense_variant', 'synonymous_variant', 'stop_gained', 'stop_lost', 'splice_region_variant', 'inframe_insertion', 'frameshift_variant']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'consequence_terms': ['3_prime_UTR_variant'],\n",
       "  'cdna_start': 2674,\n",
       "  'transcript_id': 'ENST00000325425',\n",
       "  'gene_id': 'ENSG00000162572',\n",
       "  'gene_symbol': 'SCNN1D',\n",
       "  'impact': 'MODIFIER',\n",
       "  'cdna_end': 2674,\n",
       "  'biotype': 'protein_coding',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'strand': 1,\n",
       "  'go': [{'go_term': 'GO:0005272', 'description': 'sodium_channel_activity'},\n",
       "   {'go_term': 'GO:0005515', 'description': 'protein_binding'},\n",
       "   {'go_term': 'GO:0005886', 'description': 'plasma_membrane'},\n",
       "   {'description': 'sodium_ion_transport', 'go_term': 'GO:0006814'},\n",
       "   {'description': 'ligand-gated_sodium_channel_activity',\n",
       "    'go_term': 'GO:0015280'},\n",
       "   {'go_term': 'GO:0016020', 'description': 'membrane'},\n",
       "   {'go_term': 'GO:0016021', 'description': 'integral_component_of_membrane'},\n",
       "   {'description': 'ion_transmembrane_transport', 'go_term': 'GO:0034220'},\n",
       "   {'go_term': 'GO:0050896', 'description': 'response_to_stimulus'},\n",
       "   {'go_term': 'GO:0050909', 'description': 'sensory_perception_of_taste'},\n",
       "   {'go_term': 'GO:0055085', 'description': 'transmembrane_transport'}],\n",
       "  'hgnc_id': 10601,\n",
       "  'variant_allele': 'A'},\n",
       " {'hgnc_id': 10601,\n",
       "  'variant_allele': 'A',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'go': [{'description': 'sodium_channel_activity', 'go_term': 'GO:0005272'},\n",
       "   {'description': 'protein_binding', 'go_term': 'GO:0005515'},\n",
       "   {'description': 'plasma_membrane', 'go_term': 'GO:0005886'},\n",
       "   {'description': 'sodium_ion_transport', 'go_term': 'GO:0006814'},\n",
       "   {'go_term': 'GO:0015280',\n",
       "    'description': 'ligand-gated_sodium_channel_activity'},\n",
       "   {'go_term': 'GO:0016020', 'description': 'membrane'},\n",
       "   {'description': 'integral_component_of_membrane', 'go_term': 'GO:0016021'},\n",
       "   {'go_term': 'GO:0034220', 'description': 'ion_transmembrane_transport'},\n",
       "   {'description': 'response_to_stimulus', 'go_term': 'GO:0050896'},\n",
       "   {'description': 'sensory_perception_of_taste', 'go_term': 'GO:0050909'},\n",
       "   {'description': 'transmembrane_transport', 'go_term': 'GO:0055085'}],\n",
       "  'strand': 1,\n",
       "  'gene_id': 'ENSG00000162572',\n",
       "  'gene_symbol': 'SCNN1D',\n",
       "  'cdna_end': 3470,\n",
       "  'impact': 'MODIFIER',\n",
       "  'biotype': 'protein_coding',\n",
       "  'consequence_terms': ['3_prime_UTR_variant'],\n",
       "  'transcript_id': 'ENST00000338555',\n",
       "  'cdna_start': 3470},\n",
       " {'variant_allele': 'A',\n",
       "  'hgnc_id': 16754,\n",
       "  'go': [{'description': 'protein_binding', 'go_term': 'GO:0005515'},\n",
       "   {'description': 'ARF_GTPase_activator_activity', 'go_term': 'GO:0008060'},\n",
       "   {'go_term': 'GO:0008270', 'description': 'zinc_ion_binding'},\n",
       "   {'go_term': 'GO:0032312',\n",
       "    'description': 'regulation_of_ARF_GTPase_activity'}],\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'strand': -1,\n",
       "  'impact': 'MODIFIER',\n",
       "  'gene_symbol': 'ACAP3',\n",
       "  'biotype': 'protein_coding',\n",
       "  'gene_id': 'ENSG00000131584',\n",
       "  'distance': 365,\n",
       "  'transcript_id': 'ENST00000353662',\n",
       "  'consequence_terms': ['downstream_gene_variant']},\n",
       " {'biotype': 'protein_coding',\n",
       "  'gene_symbol': 'ACAP3',\n",
       "  'impact': 'MODIFIER',\n",
       "  'gene_id': 'ENSG00000131584',\n",
       "  'distance': 357,\n",
       "  'transcript_id': 'ENST00000354700',\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'variant_allele': 'A',\n",
       "  'hgnc_id': 16754,\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'strand': -1,\n",
       "  'go': [{'go_term': 'GO:0005515', 'description': 'protein_binding'},\n",
       "   {'go_term': 'GO:0008060', 'description': 'ARF_GTPase_activator_activity'},\n",
       "   {'go_term': 'GO:0008270', 'description': 'zinc_ion_binding'},\n",
       "   {'description': 'regulation_of_ARF_GTPase_activity',\n",
       "    'go_term': 'GO:0032312'}]},\n",
       " {'gene_symbol_source': 'HGNC',\n",
       "  'strand': -1,\n",
       "  'variant_allele': 'A',\n",
       "  'hgnc_id': 16754,\n",
       "  'transcript_id': 'ENST00000379037',\n",
       "  'distance': 4417,\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'gene_symbol': 'ACAP3',\n",
       "  'impact': 'MODIFIER',\n",
       "  'biotype': 'processed_transcript',\n",
       "  'gene_id': 'ENSG00000131584'},\n",
       " {'strand': 1,\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'go': [{'description': 'sodium_channel_activity', 'go_term': 'GO:0005272'},\n",
       "   {'go_term': 'GO:0006814', 'description': 'sodium_ion_transport'},\n",
       "   {'go_term': 'GO:0016020', 'description': 'membrane'}],\n",
       "  'flags': ['cds_start_NF', 'cds_end_NF'],\n",
       "  'hgnc_id': 10601,\n",
       "  'variant_allele': 'A',\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'transcript_id': 'ENST00000379099',\n",
       "  'distance': 1076,\n",
       "  'gene_id': 'ENSG00000162572',\n",
       "  'gene_symbol': 'SCNN1D',\n",
       "  'biotype': 'protein_coding',\n",
       "  'impact': 'MODIFIER'},\n",
       " {'hgnc_id': 10601,\n",
       "  'variant_allele': 'A',\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'strand': 1,\n",
       "  'gene_id': 'ENSG00000162572',\n",
       "  'impact': 'MODIFIER',\n",
       "  'gene_symbol': 'SCNN1D',\n",
       "  'biotype': 'nonsense_mediated_decay',\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'transcript_id': 'ENST00000379101',\n",
       "  'distance': 646},\n",
       " {'gene_id': 'ENSG00000162572',\n",
       "  'gene_symbol': 'SCNN1D',\n",
       "  'impact': 'MODIFIER',\n",
       "  'cdna_end': 3044,\n",
       "  'biotype': 'protein_coding',\n",
       "  'transcript_id': 'ENST00000379116',\n",
       "  'cdna_start': 3044,\n",
       "  'consequence_terms': ['3_prime_UTR_variant'],\n",
       "  'hgnc_id': 10601,\n",
       "  'variant_allele': 'A',\n",
       "  'strand': 1,\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'go': [{'go_term': 'GO:0005272', 'description': 'sodium_channel_activity'},\n",
       "   {'go_term': 'GO:0006814', 'description': 'sodium_ion_transport'},\n",
       "   {'description': 'ligand-gated_sodium_channel_activity',\n",
       "    'go_term': 'GO:0015280'},\n",
       "   {'description': 'membrane', 'go_term': 'GO:0016020'},\n",
       "   {'description': 'integral_component_of_membrane',\n",
       "    'go_term': 'GO:0016021'}]},\n",
       " {'hgnc_id': 10601,\n",
       "  'variant_allele': 'A',\n",
       "  'go': [{'go_term': 'GO:0005272', 'description': 'sodium_channel_activity'},\n",
       "   {'go_term': 'GO:0005515', 'description': 'protein_binding'},\n",
       "   {'go_term': 'GO:0005886', 'description': 'plasma_membrane'},\n",
       "   {'description': 'sodium_ion_transport', 'go_term': 'GO:0006814'},\n",
       "   {'go_term': 'GO:0015280',\n",
       "    'description': 'ligand-gated_sodium_channel_activity'},\n",
       "   {'go_term': 'GO:0016020', 'description': 'membrane'},\n",
       "   {'go_term': 'GO:0016021', 'description': 'integral_component_of_membrane'},\n",
       "   {'go_term': 'GO:0034220', 'description': 'ion_transmembrane_transport'},\n",
       "   {'description': 'response_to_stimulus', 'go_term': 'GO:0050896'},\n",
       "   {'go_term': 'GO:0050909', 'description': 'sensory_perception_of_taste'},\n",
       "   {'description': 'transmembrane_transport', 'go_term': 'GO:0055085'}],\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'strand': 1,\n",
       "  'gene_id': 'ENSG00000162572',\n",
       "  'cdna_end': 2493,\n",
       "  'gene_symbol': 'SCNN1D',\n",
       "  'biotype': 'protein_coding',\n",
       "  'impact': 'MODIFIER',\n",
       "  'transcript_id': 'ENST00000400928',\n",
       "  'cdna_start': 2493,\n",
       "  'consequence_terms': ['3_prime_UTR_variant']},\n",
       " {'variant_allele': 'A',\n",
       "  'hgnc_id': 16754,\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'strand': -1,\n",
       "  'gene_symbol': 'ACAP3',\n",
       "  'biotype': 'retained_intron',\n",
       "  'impact': 'MODIFIER',\n",
       "  'gene_id': 'ENSG00000131584',\n",
       "  'distance': 851,\n",
       "  'transcript_id': 'ENST00000467278',\n",
       "  'consequence_terms': ['downstream_gene_variant']},\n",
       " {'variant_allele': 'A',\n",
       "  'hgnc_id': 16754,\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'strand': -1,\n",
       "  'gene_symbol': 'ACAP3',\n",
       "  'biotype': 'retained_intron',\n",
       "  'impact': 'MODIFIER',\n",
       "  'gene_id': 'ENSG00000131584',\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'distance': 2480,\n",
       "  'transcript_id': 'ENST00000470659'},\n",
       " {'variant_allele': 'A',\n",
       "  'hgnc_id': 16754,\n",
       "  'strand': -1,\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'biotype': 'retained_intron',\n",
       "  'gene_symbol': 'ACAP3',\n",
       "  'impact': 'MODIFIER',\n",
       "  'gene_id': 'ENSG00000131584',\n",
       "  'transcript_id': 'ENST00000476572',\n",
       "  'distance': 4447,\n",
       "  'consequence_terms': ['downstream_gene_variant']},\n",
       " {'distance': 371,\n",
       "  'transcript_id': 'ENST00000492936',\n",
       "  'consequence_terms': ['downstream_gene_variant'],\n",
       "  'gene_symbol': 'ACAP3',\n",
       "  'impact': 'MODIFIER',\n",
       "  'biotype': 'retained_intron',\n",
       "  'gene_id': 'ENSG00000131584',\n",
       "  'strand': -1,\n",
       "  'gene_symbol_source': 'HGNC',\n",
       "  'variant_allele': 'A',\n",
       "  'hgnc_id': 16754}]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1['transcript_consequences'][1]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ensemble",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
