{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'scripts.assembly_converter'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[1], line 6\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mintervaltree\u001b[39;00m \u001b[39mimport\u001b[39;00m Interval, IntervalTree\n\u001b[0;32m      5\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mmultiprocessing\u001b[39;00m\n\u001b[1;32m----> 6\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mscripts\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39massembly_converter\u001b[39;00m \u001b[39mimport\u001b[39;00m convert_assembly_hg19_to_hg38\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'scripts.assembly_converter'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from anytree import Node, RenderTree, AsciiStyle, LevelGroupOrderIter, LevelOrderGroupIter, PreOrderIter\n",
    "from intervaltree import Interval, IntervalTree\n",
    "import multiprocessing\n",
    "from scripts.assembly_converter import convert_assembly_hg19_to_hg38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = convert_assembly_hg19_to_hg38(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>grch_37</th>\n",
       "      <th>orientation</th>\n",
       "      <th>grch37_start</th>\n",
       "      <th>grch37_chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16</td>\n",
       "      <td>289439</td>\n",
       "      <td>289439</td>\n",
       "      <td>chr16:339440-339439</td>\n",
       "      <td>1</td>\n",
       "      <td>339439</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>289608</td>\n",
       "      <td>289608</td>\n",
       "      <td>chr16:339609-339608</td>\n",
       "      <td>1</td>\n",
       "      <td>339608</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16</td>\n",
       "      <td>297721</td>\n",
       "      <td>297721</td>\n",
       "      <td>chr16:347722-347721</td>\n",
       "      <td>1</td>\n",
       "      <td>347721</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16</td>\n",
       "      <td>310070</td>\n",
       "      <td>310070</td>\n",
       "      <td>chr16:360071-360070</td>\n",
       "      <td>1</td>\n",
       "      <td>360070</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>346146</td>\n",
       "      <td>346146</td>\n",
       "      <td>chr16:396147-396146</td>\n",
       "      <td>1</td>\n",
       "      <td>396146</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149</th>\n",
       "      <td>6</td>\n",
       "      <td>53161598</td>\n",
       "      <td>53161598</td>\n",
       "      <td>chr6:53026397-53026396</td>\n",
       "      <td>1</td>\n",
       "      <td>53026396</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1150</th>\n",
       "      <td>2</td>\n",
       "      <td>219212438</td>\n",
       "      <td>219212438</td>\n",
       "      <td>chr2:220077161-220077160</td>\n",
       "      <td>1</td>\n",
       "      <td>220077160</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1151</th>\n",
       "      <td>16</td>\n",
       "      <td>31459888</td>\n",
       "      <td>31459888</td>\n",
       "      <td>chr16:31471210-31471209</td>\n",
       "      <td>1</td>\n",
       "      <td>31471209</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1152</th>\n",
       "      <td>18</td>\n",
       "      <td>61917060</td>\n",
       "      <td>61917060</td>\n",
       "      <td>chr18:59584294-59584293</td>\n",
       "      <td>1</td>\n",
       "      <td>59584293</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1153</th>\n",
       "      <td>19</td>\n",
       "      <td>45473565</td>\n",
       "      <td>45473565</td>\n",
       "      <td>chr19:45976824-45976823</td>\n",
       "      <td>1</td>\n",
       "      <td>45976823</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1154 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr      start        end                   grch_37  orientation  \\\n",
       "0     16     289439     289439       chr16:339440-339439            1   \n",
       "1     16     289608     289608       chr16:339609-339608            1   \n",
       "2     16     297721     297721       chr16:347722-347721            1   \n",
       "3     16     310070     310070       chr16:360071-360070            1   \n",
       "4     16     346146     346146       chr16:396147-396146            1   \n",
       "...   ..        ...        ...                       ...          ...   \n",
       "1149   6   53161598   53161598    chr6:53026397-53026396            1   \n",
       "1150   2  219212438  219212438  chr2:220077161-220077160            1   \n",
       "1151  16   31459888   31459888   chr16:31471210-31471209            1   \n",
       "1152  18   61917060   61917060   chr18:59584294-59584293            1   \n",
       "1153  19   45473565   45473565   chr19:45976824-45976823            1   \n",
       "\n",
       "     grch37_start grch37_chr  \n",
       "0          339439         16  \n",
       "1          339608         16  \n",
       "2          347721         16  \n",
       "3          360070         16  \n",
       "4          396146         16  \n",
       "...           ...        ...  \n",
       "1149     53026396          6  \n",
       "1150    220077160          2  \n",
       "1151     31471209         16  \n",
       "1152     59584293         18  \n",
       "1153     45976823         19  \n",
       "\n",
       "[1154 rows x 7 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"artifacts/liftover_grch38_final.bed\", sep=\"\\t\", header = None)\n",
    "df.columns = ['chr', 'start', 'end', 'grch_37', 'orientation']\n",
    "df['grch37_start'] = df['grch_37'].apply(lambda x: x.split(':')[1].split('-')[1])\n",
    "df['grch37_chr'] = df['grch_37'].apply(lambda x: x.split(':')[0])\n",
    "df['chr'] = df['chr'].apply(lambda x: x.replace('chr', ''))\n",
    "df['grch37_chr'] = df['grch37_chr'].apply(lambda x: x.replace('chr', ''))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Entrez GeneId</th>\n",
       "      <th>Genome Location</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Hallmark</th>\n",
       "      <th>Chr Band</th>\n",
       "      <th>Somatic</th>\n",
       "      <th>Germline</th>\n",
       "      <th>Tumour Types(Somatic)</th>\n",
       "      <th>...</th>\n",
       "      <th>Molecular Genetics</th>\n",
       "      <th>Role in Cancer</th>\n",
       "      <th>Mutation Types</th>\n",
       "      <th>Translocation Partner</th>\n",
       "      <th>Other Germline Mut</th>\n",
       "      <th>Other Syndrome</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>29974.0</td>\n",
       "      <td>10:50799421-50885675</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.23</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>melanoma</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF,ACF64,ACF65,APOBEC1CF,ASP,CCDS73133.1,ENSG...</td>\n",
       "      <td>10</td>\n",
       "      <td>50799421</td>\n",
       "      <td>50885675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>abl-interactor 1</td>\n",
       "      <td>10006.0</td>\n",
       "      <td>10:26746593-26860935</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>12.10</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>...</td>\n",
       "      <td>Dom</td>\n",
       "      <td>TSG, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABI-1,CCDS7150.1,E3B1,ENSG00000136754.17,NM_00...</td>\n",
       "      <td>10</td>\n",
       "      <td>26746593</td>\n",
       "      <td>26860935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene h...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>9:130713946-130885683</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>34.12</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CML, ALL, T-ALL</td>\n",
       "      <td>...</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T, Mis</td>\n",
       "      <td>BCR, ETV6, NUP214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABL,CCDS35165.1,ENSG00000097007.17,JTK7,NM_007...</td>\n",
       "      <td>9</td>\n",
       "      <td>130713946</td>\n",
       "      <td>130885683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABL2</td>\n",
       "      <td>c-abl oncogene 2, non-receptor tyrosine kinase</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1:179099327-179229601</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.20</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>...</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>ETV6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ABLL,ARG,CCDS30947.1,ENSG00000143322.19,NM_007...</td>\n",
       "      <td>1</td>\n",
       "      <td>179099327</td>\n",
       "      <td>179229601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACKR3</td>\n",
       "      <td>atypical chemokine receptor 3</td>\n",
       "      <td>57007.0</td>\n",
       "      <td>2:236569641-236582358</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>37.30</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lipoma</td>\n",
       "      <td>...</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>HMGA2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS2516.1,CMKOR1,CXCR7,ENSG00000144476.5,GPR1...</td>\n",
       "      <td>2</td>\n",
       "      <td>236569641</td>\n",
       "      <td>236582358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731</th>\n",
       "      <td>ZNF429</td>\n",
       "      <td>zinc finger protein 429</td>\n",
       "      <td>353088.0</td>\n",
       "      <td>19:21505564-21538078</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.00</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GBM</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS42537.1,ENSG00000197013.9,NM_001001415.2,N...</td>\n",
       "      <td>19</td>\n",
       "      <td>21505564</td>\n",
       "      <td>21538078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>732</th>\n",
       "      <td>ZNF479</td>\n",
       "      <td>zinc finger protein 479</td>\n",
       "      <td>90827.0</td>\n",
       "      <td>7:57119614-57139864</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.20</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lung cancer, bladder carcinoma, prostate carci...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS43590.1,ENSG00000185177.12,KR19,NM_033273....</td>\n",
       "      <td>7</td>\n",
       "      <td>57119614</td>\n",
       "      <td>57139864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>ZNF521</td>\n",
       "      <td>zinc finger protein 521</td>\n",
       "      <td>25925.0</td>\n",
       "      <td>18:25061926-25352152</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.20</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ALL</td>\n",
       "      <td>...</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>PAX5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS32806.1,EHZF,ENSG00000198795.10,Evi3,NM_01...</td>\n",
       "      <td>18</td>\n",
       "      <td>25061926</td>\n",
       "      <td>25352152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>734</th>\n",
       "      <td>ZNRF3</td>\n",
       "      <td>zinc and ring finger 3</td>\n",
       "      <td>84133.0</td>\n",
       "      <td>22:28883592-29057487</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.10</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>colorectal cancer, adrenocortical carcinoma, g...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>N, F, Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BK747E2.3,CCDS56225.1,ENSG00000183579.15,FLJ22...</td>\n",
       "      <td>22</td>\n",
       "      <td>28883592</td>\n",
       "      <td>29057487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>735</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>zinc finger (CCCH type), RNA-binding motif and...</td>\n",
       "      <td>8233.0</td>\n",
       "      <td>X:15790472-15823260</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>22.20</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MDS, CLL</td>\n",
       "      <td>...</td>\n",
       "      <td>Rec</td>\n",
       "      <td>TSG</td>\n",
       "      <td>F, S, Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS14172.1,ENSG00000169249.12,NM_005089.3,NP_...</td>\n",
       "      <td>X</td>\n",
       "      <td>15790472</td>\n",
       "      <td>15823260</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>730 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene Symbol                                               Name  \\\n",
       "0          A1CF                     APOBEC1 complementation factor   \n",
       "1          ABI1                                   abl-interactor 1   \n",
       "2          ABL1  v-abl Abelson murine leukemia viral oncogene h...   \n",
       "3          ABL2     c-abl oncogene 2, non-receptor tyrosine kinase   \n",
       "4         ACKR3                      atypical chemokine receptor 3   \n",
       "..          ...                                                ...   \n",
       "731      ZNF429                            zinc finger protein 429   \n",
       "732      ZNF479                            zinc finger protein 479   \n",
       "733      ZNF521                            zinc finger protein 521   \n",
       "734       ZNRF3                             zinc and ring finger 3   \n",
       "735       ZRSR2  zinc finger (CCCH type), RNA-binding motif and...   \n",
       "\n",
       "     Entrez GeneId        Genome Location  Tier Hallmark  Chr Band Somatic  \\\n",
       "0          29974.0   10:50799421-50885675     2      NaN     11.23     yes   \n",
       "1          10006.0   10:26746593-26860935     1      Yes     12.10     yes   \n",
       "2             25.0  9:130713946-130885683     1      Yes     34.12     yes   \n",
       "3             27.0  1:179099327-179229601     1      NaN     25.20     yes   \n",
       "4          57007.0  2:236569641-236582358     1      Yes     37.30     yes   \n",
       "..             ...                    ...   ...      ...       ...     ...   \n",
       "731       353088.0   19:21505564-21538078     2      NaN     12.00     yes   \n",
       "732        90827.0    7:57119614-57139864     2      NaN     11.20     yes   \n",
       "733        25925.0   18:25061926-25352152     1      NaN     11.20     yes   \n",
       "734        84133.0   22:28883592-29057487     2      NaN     12.10     yes   \n",
       "735         8233.0    X:15790472-15823260     1      Yes     22.20     yes   \n",
       "\n",
       "    Germline                              Tumour Types(Somatic)  ...  \\\n",
       "0        NaN                                           melanoma  ...   \n",
       "1        NaN                                                AML  ...   \n",
       "2        NaN                                    CML, ALL, T-ALL  ...   \n",
       "3        NaN                                                AML  ...   \n",
       "4        NaN                                             lipoma  ...   \n",
       "..       ...                                                ...  ...   \n",
       "731      NaN                                                GBM  ...   \n",
       "732      NaN  lung cancer, bladder carcinoma, prostate carci...  ...   \n",
       "733      NaN                                                ALL  ...   \n",
       "734      NaN  colorectal cancer, adrenocortical carcinoma, g...  ...   \n",
       "735      NaN                                           MDS, CLL  ...   \n",
       "\n",
       "    Molecular Genetics    Role in Cancer Mutation Types Translocation Partner  \\\n",
       "0                  NaN          oncogene            Mis                   NaN   \n",
       "1                  Dom       TSG, fusion              T                 KMT2A   \n",
       "2                  Dom  oncogene, fusion         T, Mis     BCR, ETV6, NUP214   \n",
       "3                  Dom  oncogene, fusion              T                  ETV6   \n",
       "4                  Dom  oncogene, fusion              T                 HMGA2   \n",
       "..                 ...               ...            ...                   ...   \n",
       "731                NaN               NaN            Mis                   NaN   \n",
       "732                NaN               NaN            Mis                   NaN   \n",
       "733                Dom  oncogene, fusion              T                  PAX5   \n",
       "734                NaN               TSG      N, F, Mis                   NaN   \n",
       "735                Rec               TSG      F, S, Mis                   NaN   \n",
       "\n",
       "    Other Germline Mut Other Syndrome  \\\n",
       "0                  NaN            NaN   \n",
       "1                  NaN            NaN   \n",
       "2                  NaN            NaN   \n",
       "3                  NaN            NaN   \n",
       "4                  NaN            NaN   \n",
       "..                 ...            ...   \n",
       "731                NaN            NaN   \n",
       "732                NaN            NaN   \n",
       "733                NaN            NaN   \n",
       "734                NaN            NaN   \n",
       "735                NaN            NaN   \n",
       "\n",
       "                                              Synonyms chr      start  \\\n",
       "0    ACF,ACF64,ACF65,APOBEC1CF,ASP,CCDS73133.1,ENSG...  10   50799421   \n",
       "1    ABI-1,CCDS7150.1,E3B1,ENSG00000136754.17,NM_00...  10   26746593   \n",
       "2    ABL,CCDS35165.1,ENSG00000097007.17,JTK7,NM_007...   9  130713946   \n",
       "3    ABLL,ARG,CCDS30947.1,ENSG00000143322.19,NM_007...   1  179099327   \n",
       "4    CCDS2516.1,CMKOR1,CXCR7,ENSG00000144476.5,GPR1...   2  236569641   \n",
       "..                                                 ...  ..        ...   \n",
       "731  CCDS42537.1,ENSG00000197013.9,NM_001001415.2,N...  19   21505564   \n",
       "732  CCDS43590.1,ENSG00000185177.12,KR19,NM_033273....   7   57119614   \n",
       "733  CCDS32806.1,EHZF,ENSG00000198795.10,Evi3,NM_01...  18   25061926   \n",
       "734  BK747E2.3,CCDS56225.1,ENSG00000183579.15,FLJ22...  22   28883592   \n",
       "735  CCDS14172.1,ENSG00000169249.12,NM_005089.3,NP_...   X   15790472   \n",
       "\n",
       "           end  \n",
       "0     50885675  \n",
       "1     26860935  \n",
       "2    130885683  \n",
       "3    179229601  \n",
       "4    236582358  \n",
       "..         ...  \n",
       "731   21538078  \n",
       "732   57139864  \n",
       "733   25352152  \n",
       "734   29057487  \n",
       "735   15823260  \n",
       "\n",
       "[730 rows x 23 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cosmic = pd.read_csv('../data/COSMIC_driver_genes.csv')\n",
    "cosmic = cosmic[~cosmic['Genome Location'].str.contains(':-')]\n",
    "cosmic['chr'] = cosmic['Genome Location'].str.split(':').map(lambda x: x[0])\n",
    "cosmic['start'] = cosmic['Genome Location'].str.split(':').map(lambda x: x[1]).str.split('-').map(lambda x: int(x[0]))\n",
    "cosmic['end'] = cosmic['Genome Location'].str.split(':').map(lambda x: x[1]).str.split('-').map(lambda x: int(x[1]))\n",
    "cosmic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "cgc_tree = IntervalTree()\n",
    "\n",
    "for index, row in cosmic.iterrows():\n",
    "    cgc_tree.add(Interval(row['start'], row['end'], (row['chr'], row['Gene Symbol'], '0kb')))\n",
    "\n",
    "    cgc_tree.add(Interval(row['start']-2000, row['start'], (row['chr'], row['Gene Symbol'], '2kb upstream')))\n",
    "    cgc_tree.add(Interval(row['start']-10000, row['start']-2000, (row['chr'], row['Gene Symbol'], '10kb upstream')))\n",
    "    cgc_tree.add(Interval(row['start']-100000, row['start']-10000, (row['chr'], row['Gene Symbol'], '100kb upstream')))\n",
    "\n",
    "    cgc_tree.add(Interval(row['end'], row['end']+2000, (row['chr'], row['Gene Symbol'], '2kb downstream')))\n",
    "    cgc_tree.add(Interval(row['end']+2000, row['end']+10000, (row['chr'], row['Gene Symbol'], '10kb downstream')))\n",
    "    cgc_tree.add(Interval(row['end']+10000, row['end']+100000, (row['chr'], row['Gene Symbol'], '100kb downstream')))\n",
    "\n",
    "df_tree = []\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    df_tree.append(Node(name = row['chr'] + ':' + str(row['start']) + '-' + str(row['end']), chr = row['chr'], start = row['start'], end = row['end']))\n",
    "\n",
    "df['interactions'] = 0\n",
    "df['interaction_gene'] = np.nan\n",
    "df['interaction_info'] = np.nan\n",
    "\n",
    "for index in range(len(df_tree)):\n",
    "    child_list = [node for node in LevelOrderGroupIter(df_tree[index])][0]\n",
    "    for node in child_list:\n",
    "        found_list = list(set(list(cgc_tree.overlap(node.start, node.end)) + list(cgc_tree.at(node.start)) + list(cgc_tree.at(node.end))))  # use its coordinates to find interactions in the IntervalTree\n",
    "        listindex = 0\n",
    "        while listindex < len(found_list) and found_list[listindex].data[0] != node.chr:\n",
    "            listindex += 1\n",
    "        # print(len(found_list), listindex, node.chr, node.start, node.end)\n",
    "        # print(found_list)\n",
    "        if listindex < len(found_list):\n",
    "            children_left = True\n",
    "            found_interaction = found_list[listindex].data\n",
    "            Node(name = found_interaction[0] + ':' +  str(found_list[listindex].begin) + '-' +  str(found_list[listindex].end), chr = found_interaction[0], start = found_list[listindex].begin, end = found_list[listindex].end, checked = 0, parent = node)\n",
    "            # print(\"Added child to:\", index)\n",
    "            df.at[index, 'interactions'] = len(df_tree[index].descendants)\n",
    "            df.at[index, 'interaction_gene'] = found_interaction[1]\n",
    "            df.at[index, 'interaction_info'] = found_interaction[2]\n",
    "    # if index == 19:\n",
    "    #         break\n",
    "\n",
    "df['grch37_start'] = pd.to_numeric(df['grch37_start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>grch_37</th>\n",
       "      <th>orientation</th>\n",
       "      <th>grch37_start</th>\n",
       "      <th>grch37_chr</th>\n",
       "      <th>interactions</th>\n",
       "      <th>interaction_gene</th>\n",
       "      <th>interaction_info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16</td>\n",
       "      <td>289439</td>\n",
       "      <td>289439</td>\n",
       "      <td>chr16:339440-339439</td>\n",
       "      <td>1</td>\n",
       "      <td>339439</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>AXIN1</td>\n",
       "      <td>0kb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>289608</td>\n",
       "      <td>289608</td>\n",
       "      <td>chr16:339609-339608</td>\n",
       "      <td>1</td>\n",
       "      <td>339608</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>AXIN1</td>\n",
       "      <td>0kb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16</td>\n",
       "      <td>297721</td>\n",
       "      <td>297721</td>\n",
       "      <td>chr16:347722-347721</td>\n",
       "      <td>1</td>\n",
       "      <td>347721</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>AXIN1</td>\n",
       "      <td>0kb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16</td>\n",
       "      <td>310070</td>\n",
       "      <td>310070</td>\n",
       "      <td>chr16:360071-360070</td>\n",
       "      <td>1</td>\n",
       "      <td>360070</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>AXIN1</td>\n",
       "      <td>0kb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>346146</td>\n",
       "      <td>346146</td>\n",
       "      <td>chr16:396147-396146</td>\n",
       "      <td>1</td>\n",
       "      <td>396146</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>AXIN1</td>\n",
       "      <td>0kb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1101</th>\n",
       "      <td>16</td>\n",
       "      <td>2254001</td>\n",
       "      <td>2254001</td>\n",
       "      <td>chr16:2304003-2304002</td>\n",
       "      <td>1</td>\n",
       "      <td>2304002</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>TRAF7</td>\n",
       "      <td>100kb downstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1119</th>\n",
       "      <td>1</td>\n",
       "      <td>156738411</td>\n",
       "      <td>156738411</td>\n",
       "      <td>chr1:156708204-156708203</td>\n",
       "      <td>1</td>\n",
       "      <td>156708203</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>PRCC</td>\n",
       "      <td>100kb upstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1121</th>\n",
       "      <td>12</td>\n",
       "      <td>57125241</td>\n",
       "      <td>57125241</td>\n",
       "      <td>chr12:57519025-57519024</td>\n",
       "      <td>1</td>\n",
       "      <td>57519024</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>STAT6</td>\n",
       "      <td>100kb downstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>10</td>\n",
       "      <td>102470972</td>\n",
       "      <td>102470972</td>\n",
       "      <td>chr10:104230730-104230729</td>\n",
       "      <td>1</td>\n",
       "      <td>104230729</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>SUFU</td>\n",
       "      <td>100kb upstream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1140</th>\n",
       "      <td>X</td>\n",
       "      <td>101356185</td>\n",
       "      <td>101356185</td>\n",
       "      <td>chrX:100611174-100611173</td>\n",
       "      <td>1</td>\n",
       "      <td>100611173</td>\n",
       "      <td>X</td>\n",
       "      <td>1</td>\n",
       "      <td>BTK</td>\n",
       "      <td>0kb</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>386 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr      start        end                    grch_37  orientation  \\\n",
       "0     16     289439     289439        chr16:339440-339439            1   \n",
       "1     16     289608     289608        chr16:339609-339608            1   \n",
       "2     16     297721     297721        chr16:347722-347721            1   \n",
       "3     16     310070     310070        chr16:360071-360070            1   \n",
       "4     16     346146     346146        chr16:396147-396146            1   \n",
       "...   ..        ...        ...                        ...          ...   \n",
       "1101  16    2254001    2254001      chr16:2304003-2304002            1   \n",
       "1119   1  156738411  156738411   chr1:156708204-156708203            1   \n",
       "1121  12   57125241   57125241    chr12:57519025-57519024            1   \n",
       "1133  10  102470972  102470972  chr10:104230730-104230729            1   \n",
       "1140   X  101356185  101356185   chrX:100611174-100611173            1   \n",
       "\n",
       "      grch37_start grch37_chr  interactions interaction_gene  interaction_info  \n",
       "0           339439         16             1            AXIN1               0kb  \n",
       "1           339608         16             1            AXIN1               0kb  \n",
       "2           347721         16             1            AXIN1               0kb  \n",
       "3           360070         16             1            AXIN1               0kb  \n",
       "4           396146         16             1            AXIN1               0kb  \n",
       "...            ...        ...           ...              ...               ...  \n",
       "1101       2304002         16             1            TRAF7  100kb downstream  \n",
       "1119     156708203          1             1             PRCC    100kb upstream  \n",
       "1121      57519024         12             1            STAT6  100kb downstream  \n",
       "1133     104230729         10             1             SUFU    100kb upstream  \n",
       "1140     100611173          X             1              BTK               0kb  \n",
       "\n",
       "[386 rows x 10 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['interactions'] > 0].drop_duplicates(keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop_duplicates(keep='first', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_pickle('artifacts/gencode_interactions.pickle', compression='infer', protocol=5, storage_options=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TP53      64\n",
       "TERT      40\n",
       "RNF43     18\n",
       "FOXA1     11\n",
       "RB1       11\n",
       "          ..\n",
       "PMS2       1\n",
       "ESR1       1\n",
       "KMT2C      1\n",
       "ACVR2A     1\n",
       "BTK        1\n",
       "Name: interaction_gene, Length: 153, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['interaction_gene'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0kb                 301\n",
       "2kb downstream       49\n",
       "100kb upstream       48\n",
       "100kb downstream     47\n",
       "10kb downstream      10\n",
       "10kb upstream         2\n",
       "2kb upstream          1\n",
       "Name: interaction_info, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['interaction_info'].value_counts()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "be43cc5c71a68f9f019d1dc826cd47f0aea8eff41b0f1d32dd7c3f8c76ea0dfa"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
